{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Eugenio Perez Molphe Montoya ####\n",
    "#### 17.06.2024 ####\n",
    "#### Create a file of taxonomy_identity_numbers for the Salmonella species ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Import packages ###\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Remove the duplicates from the metadata ####\n",
    "\n",
    "# I have no writing access to Nico's directory, so I will copy the metadata file to my directory but this new metadata will have no duplicates\n",
    "out = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/salmonella_metadata.txt'\n",
    "\n",
    "# Read the metadata file\n",
    "pathMetadata = '/mnt/mnemo6/nnaepf/for_eugenio/data/salmonella_genomes/salmonella_metadata.txt'\n",
    "metadata = pd.read_csv(pathMetadata)\n",
    "\n",
    "# Read the list of duplicates\n",
    "pathDuplicates = '/mnt/mnemo6/nnaepf/for_eugenio/data/salmonella_genomes/duplicate_genomes.txt'\n",
    "with open(pathDuplicates, 'r') as f:\n",
    "    duplicates = f.readlines()\n",
    "duplicates = [x.strip() for x in duplicates]\n",
    "\n",
    "# Remove the duplicates\n",
    "metadata = metadata[~metadata['accession'].isin(duplicates)]\n",
    "\n",
    "# Save the metadata without duplicates\n",
    "metadata.to_csv(out, sep=',', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create a dictionary with the following structure: ###\n",
    "# {header_code: NCBI_taxonomic_code}\n",
    "\n",
    "# Sample data: list of FASTA file names\n",
    "directory = '/mnt/mnemo6/nnaepf/for_eugenio/data/salmonella_genomes/ncbi_dataset/genomes'\n",
    "\n",
    "allFastaFiles = os.listdir(directory)\n",
    "\n",
    "# Create an empty dictionary to hold the mappings\n",
    "headerToNCBIcode = {}\n",
    "\n",
    "# Iterate through each FASTA file\n",
    "for file_name in allFastaFiles:\n",
    "    # Check if file exists to avoid errors\n",
    "    with open(directory + '/' + file_name, 'r') as file:\n",
    "        # Let's extract the codes in the headers of the fasta file\n",
    "        lines = file.readlines()\n",
    "        headersFastaFile = [line for line in lines if line.startswith('>')]\n",
    "        headersFastaFile = [header[1:].strip() for header in headersFastaFile]\n",
    "        headersFastaFile = [header.split(' ')[0] for header in headersFastaFile]\n",
    "        # Add the mapping to the dictionary\n",
    "        for header in headersFastaFile:\n",
    "            headerToNCBIcode[header] = file_name\n",
    "\n",
    "# Only get the NCBI taxonomic code for the file names\n",
    "for key in headerToNCBIcode:\n",
    "    headerToNCBIcode[key] = headerToNCBIcode[key].split('_')[0] + '_' + headerToNCBIcode[key].split('_')[1]\n",
    "\n",
    "# Print the results\n",
    "print(headerToNCBIcode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create a file with the header code and NCBI taxonomic code for the Salmonella species ##\n",
    "# Save the dictionary to a file\n",
    "out = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Salmonella_header_NCBIcode.csv'\n",
    "with open(out, 'w') as f:\n",
    "    for key in headerToNCBIcode:\n",
    "        f.write(key + ',' + headerToNCBIcode[key] + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## In case I have already the header code and NCBI code file, I can read it ##\n",
    "# Read the file\n",
    "path = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Salmonella_header_NCBIcode.csv'\n",
    "headerToNCBIcode = {}\n",
    "with open(path, 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    for line in lines:\n",
    "        header, NCBIcode = line.strip().split(',')\n",
    "        headerToNCBIcode[header] = NCBIcode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read the fasta file of the Salmonella species and extract the headers ###\n",
    "\n",
    "# The path of the fasta file\n",
    "pathFasta = '/mnt/mnemo5/eugenio/IroN_project/Files/04_Blast_annotations/salmonella_all_hits.fa'\n",
    "pathListofTaxonomyCodes = '/mnt/mnemo5/eugenio/IroN_project/Files/04_Blast_annotations/salmonella_blasted_sequences.ncbi_codes.txt' # Output file\n",
    "\n",
    "# Open the file\n",
    "with open(pathFasta, 'r') as file:\n",
    "    # Read the file\n",
    "    lines = file.readlines()\n",
    "    # Extract the headers\n",
    "    headersFastaFile = [line for line in lines if line.startswith('>')]\n",
    "    # Remove the character '>' and the newline character\n",
    "    headersFastaFile = [header[1:].strip() for header in headersFastaFile]\n",
    "    # Remove everything after the first space\n",
    "    headersFastaFile = [header.split(' ')[0] for header in headersFastaFile]\n",
    "    # Remove the _upstream_seq thing\n",
    "    headersFastaFile = [header.strip('_upstream_seq') for header in headersFastaFile]\n",
    "\n",
    "# Time to replace the headers with the taxonomic code\n",
    "\n",
    "salmonellaAccesions = []\n",
    "for i in headersFastaFile:\n",
    "    if i in headerToNCBIcode:\n",
    "        salmonellaAccesions.append(headerToNCBIcode[i])\n",
    "\n",
    "print(len(salmonellaAccesions))\n",
    "\n",
    "# Save it as a list and save the list as a file\n",
    "\n",
    "with open(pathListofTaxonomyCodes, 'w') as file:\n",
    "    for header in salmonellaAccesions:\n",
    "        file.write(header + '\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I already run this code, so everything is ready for the following script: get_taxonomic_information.ipynb\n",
    "9.07.2024"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 57\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m strings\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# Now let's get the assembly codes for the headers that are not assembly codes\u001b[39;00m\n\u001b[0;32m---> 57\u001b[0m updatedCodes \u001b[38;5;241m=\u001b[39m \u001b[43mreplace_with_fasta_name\u001b[49m\u001b[43m(\u001b[49m\u001b[43mheadersFastaFile\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpathDirectory\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;66;03m# Save the updated codes\u001b[39;00m\n\u001b[1;32m     60\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(pathListofTaxonomyCodes, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mw\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m file:\n",
      "Cell \u001b[0;32mIn[11], line 48\u001b[0m, in \u001b[0;36mreplace_with_fasta_name\u001b[0;34m(strings, fasta_directory)\u001b[0m\n\u001b[1;32m     46\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m fasta_file \u001b[38;5;129;01min\u001b[39;00m fasta_files:\n\u001b[1;32m     47\u001b[0m     fasta_path \u001b[38;5;241m=\u001b[39m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(fasta_directory, fasta_file)\n\u001b[0;32m---> 48\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mfind_string_in_fasta\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfasta_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43ms\u001b[49m\u001b[43m)\u001b[49m:\n\u001b[1;32m     49\u001b[0m         strings[i] \u001b[38;5;241m=\u001b[39m fasta_file  \u001b[38;5;66;03m# Replace header with FASTA file name\u001b[39;00m\n\u001b[1;32m     50\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mReplaced \u001b[39m\u001b[38;5;132;01m{\u001b[39;00ms\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m with \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfasta_file\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "Cell \u001b[0;32mIn[11], line 35\u001b[0m, in \u001b[0;36mfind_string_in_fasta\u001b[0;34m(fasta_file, search_string)\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Searches for the presence of a specific string (header) in a FASTA file.\"\"\"\u001b[39;00m\n\u001b[1;32m     34\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(fasta_file, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[0;32m---> 35\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m line \u001b[38;5;129;01min\u001b[39;00m f:\n\u001b[1;32m     36\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m search_string \u001b[38;5;129;01min\u001b[39;00m line:\n\u001b[1;32m     37\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/lib/python3.10/codecs.py:319\u001b[0m, in \u001b[0;36mBufferedIncrementalDecoder.decode\u001b[0;34m(self, input, final)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_buffer_decode\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m, errors, final):\n\u001b[1;32m    315\u001b[0m     \u001b[38;5;66;03m# Overwrite this method in subclasses: It must decode input\u001b[39;00m\n\u001b[1;32m    316\u001b[0m     \u001b[38;5;66;03m# and return an (output, length consumed) tuple\u001b[39;00m\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m\n\u001b[0;32m--> 319\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mdecode\u001b[39m(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;28minput\u001b[39m, final\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m):\n\u001b[1;32m    320\u001b[0m     \u001b[38;5;66;03m# decode input (taking the buffer into account)\u001b[39;00m\n\u001b[1;32m    321\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuffer \u001b[38;5;241m+\u001b[39m \u001b[38;5;28minput\u001b[39m\n\u001b[1;32m    322\u001b[0m     (result, consumed) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_buffer_decode(data, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39merrors, final)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#### 15.10.2024 ####\n",
    "#### I do the same for the E. coli species ####\n",
    "\n",
    "# I'll read the NCBI codes for the E. coli species from the metadata file\n",
    "pathFasta = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Escherichia_iroN.fa'\n",
    "# To find the assembly codes for the headers that are not assembly codes, I need to search for the headers in the FASTA files\n",
    "pathDirectory = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Genomes'\n",
    "# The output file\n",
    "pathListofTaxonomyCodes = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Results/Escherichia_w_iroN.txt'\n",
    "\n",
    "# Open the file\n",
    "with open(pathFasta, 'r') as file:\n",
    "    # Read the file\n",
    "    lines = file.readlines()\n",
    "    # Extract the headers\n",
    "    headersFastaFile = [line for line in lines if line.startswith('>')]\n",
    "    # Remove the character '>' and the newline character\n",
    "    headersFastaFile = [header[1:].strip() for header in headersFastaFile]\n",
    "    # Remove everything after the first space\n",
    "    headersFastaFile = [header.split(' ')[0] for header in headersFastaFile]\n",
    "\n",
    "# Now, we are in a bit of a sticky situation: some headers are actually the name of the contig, not the assembly code, so I need to search for the contig in the fasta files\n",
    "# The functions for this thing\n",
    "def get_fasta_files(directory):\n",
    "    \"\"\"Returns a list of all FASTA files in the directory.\"\"\"\n",
    "    return [f for f in os.listdir(directory) if f.endswith('.fna') or f.endswith('.fa')]\n",
    "\n",
    "def is_gca_string(s):\n",
    "    \"\"\"Checks if a string starts with 'GCA_' indicating it is a valid assembly name.\"\"\"\n",
    "    return s.startswith('GCA_')\n",
    "\n",
    "def find_string_in_fasta(fasta_file, search_string):\n",
    "    \"\"\"Searches for the presence of a specific string (header) in a FASTA file.\"\"\"\n",
    "    with open(fasta_file, 'r') as f:\n",
    "        for line in f:\n",
    "            if search_string in line:\n",
    "                return True\n",
    "    return False\n",
    "\n",
    "def replace_with_fasta_name(strings, fasta_directory):\n",
    "    \"\"\"Replaces non-GCA strings with the names of the FASTA files they belong to.\"\"\"\n",
    "    fasta_files = get_fasta_files(fasta_directory)\n",
    "    \n",
    "    for i, s in enumerate(strings):\n",
    "        if not is_gca_string(s):  # If the string is not a GCA_ string (header of contig)\n",
    "            for fasta_file in fasta_files:\n",
    "                fasta_path = os.path.join(fasta_directory, fasta_file)\n",
    "                if find_string_in_fasta(fasta_path, s):\n",
    "                    strings[i] = fasta_file  # Replace header with FASTA file name\n",
    "                    print(f\"Replaced {s} with {fasta_file}\")\n",
    "                    break\n",
    "\n",
    "    strings = [code.split('_')[0] + '_' + code.split('_')[1] for code in strings]  # Remove the file name and keep only the assembly code\n",
    "    return strings\n",
    "\n",
    "# Now let's get the assembly codes for the headers that are not assembly codes\n",
    "updatedCodes = replace_with_fasta_name(headersFastaFile, pathDirectory)\n",
    "\n",
    "# Save the updated codes\n",
    "with open(pathListofTaxonomyCodes, 'w') as file:\n",
    "    for code in updatedCodes:\n",
    "        file.write(code + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Replaced RODF01000039 with GCA_003794675.1_PDT000395251.1_genomic.fna\n",
      "Replaced AAZTQD010000042 with GCA_017753105.1_PDT000997500.1_genomic.fna\n",
      "Replaced AASWKY010000074 with GCA_012609475.1_PDT000639540.1_genomic.fna\n",
      "Replaced AASVZO010000056 with GCA_012602615.1_PDT000644926.1_genomic.fna\n",
      "Replaced AASTQI010000050 with GCA_012567285.1_PDT000156806.2_genomic.fna\n",
      "Replaced AATLXQ010000021 with GCA_012898485.1_PDT000215551.2_genomic.fna\n",
      "Replaced DACYOK010000047 with GCA_016108845.1_PDT000847298.1_genomic.fna\n",
      "Replaced AASEDK010000025 with GCA_012359775.1_PDT000497308.1_genomic.fna\n",
      "Replaced AAZIGU010000025 with GCA_017253655.1_PDT000976836.1_genomic.fna\n",
      "Replaced AAZZEP010000040 with GCA_017924855.1_PDT001004631.1_genomic.fna\n",
      "Replaced DAEVMD010000009 with GCA_019883145.1_PDT001123324.1_genomic.fna\n",
      "Replaced DABSCJ010000037 with GCA_014485755.1_PDT000829345.1_genomic.fna\n",
      "Replaced AARMNU010000106 with GCA_012045625.1_PDT000420662.1_genomic.fna\n",
      "Replaced DADVUR010000108 with GCA_017947765.1_PDT001005121.1_genomic.fna\n",
      "Replaced DAEABG010000048 with GCA_018180115.1_PDT001013682.1_genomic.fna\n",
      "Replaced DAERAY010000047 with GCA_019002085.1_PDT001075666.1_genomic.fna\n",
      "Replaced DABDPU010000036 with GCA_012790235.1_PDT000694455.1_genomic.fna\n",
      "Replaced UGAK01000003 with GCA_900448695.1_38273_D01_genomic.fna\n",
      "Replaced AASFPT010000063 with GCA_012298145.1_PDT000548086.1_genomic.fna\n",
      "Replaced DABWYD010000033 with GCA_014822515.1_PDT000849801.1_genomic.fna\n",
      "Replaced DADPGS010000050 with GCA_017675485.1_PDT000995950.1_genomic.fna\n",
      "Replaced AASIBA010000043 with GCA_012355435.1_PDT000257710.2_genomic.fna\n",
      "Replaced NMHS01000076 with GCA_002513145.1_ASM251314v1_genomic.fna\n",
      "Replaced DABEJM010000030 with GCA_012865015.1_PDT000713797.1_genomic.fna\n",
      "Replaced AAXJXF010000072 with GCA_015621365.1_PDT000893815.1_genomic.fna\n",
      "Replaced DADZKZ010000047 with GCA_018162935.1_PDT001013215.1_genomic.fna\n",
      "Replaced DAEAPX010000123 with GCA_018258815.1_PDT001018079.1_genomic.fna\n",
      "Replaced AATOER010000070 with GCA_012989585.1_PDT000566164.1_genomic.fna\n",
      "Replaced DAEFNY010000038 with GCA_018560825.1_PDT001040013.1_genomic.fna\n",
      "Replaced JAEJRS010000057 with GCA_016387985.1_ASM1638798v1_genomic.fna\n",
      "Replaced DADZKP010000045 with GCA_018162755.1_PDT001013231.1_genomic.fna\n",
      "Replaced AASLKP010000052 with GCA_012416045.1_PDT000240670.2_genomic.fna\n",
      "Replaced DABBBR010000054 with GCA_012534495.1_PDT000619631.1_genomic.fna\n",
      "Replaced AASCHE010000053 with GCA_012279835.1_PDT000481214.1_genomic.fna\n",
      "Replaced NBJK01000003 with GCA_019297555.1_ASM1929755v1_genomic.fna\n",
      "Replaced AASPCL010000094 with GCA_012469615.1_PDT000610080.1_genomic.fna\n",
      "Replaced AAWHKR010000033 with GCA_014736815.1_PDT000844554.1_genomic.fna\n",
      "Replaced AASXNK010000036 with GCA_012625005.1_PDT000594980.1_genomic.fna\n",
      "Replaced ABCPSG010000040 with GCA_020091045.1_PDT001134887.1_genomic.fna\n",
      "Replaced AASDOY010000032 with GCA_012383135.1_PDT000505387.1_genomic.fna\n",
      "Replaced AARZWZ010000046 with GCA_012224975.1_PDT000299829.2_genomic.fna\n",
      "Replaced DABVUJ010000105 with GCA_014751945.1_PDT000845080.1_genomic.fna\n",
      "Replaced AASDVH010000041 with GCA_012358635.1_PDT000500478.1_genomic.fna\n",
      "Replaced DABQZL010000047 with GCA_014461235.1_PDT000827934.1_genomic.fna\n",
      "Replaced JAEOJK010000044 with GCA_019366045.1_ASM1936604v1_genomic.fna\n",
      "Replaced JACXNG010000207 with GCA_014990325.1_ASM1499032v1_genomic.fna\n",
      "Replaced AATOHH010000020 with GCA_012990985.1_PDT000566094.1_genomic.fna\n",
      "Replaced DADEQA010000056 with GCA_016698065.1_PDT000937428.1_genomic.fna\n",
      "Replaced JACWZV010000064 with GCA_014809755.1_ASM1480975v1_genomic.fna\n",
      "Replaced AASXFK010000070 with GCA_012621345.1_PDT000620252.1_genomic.fna\n",
      "Replaced DAEVQU010000091 with GCA_019885965.1_PDT001123185.1_genomic.fna\n",
      "Replaced AASHIP010000050 with GCA_012343035.1_PDT000264624.2_genomic.fna\n",
      "Replaced AASGGJ010000092 with GCA_012314505.1_PDT000272549.2_genomic.fna\n",
      "Replaced DADIGG010000065 with GCA_017047755.1_PDT000953670.1_genomic.fna\n",
      "Replaced AASUGT010000047 with GCA_012576525.1_PDT000155470.2_genomic.fna\n",
      "Replaced AAYZDP010000079 with GCA_017031395.1_PDT000954877.1_genomic.fna\n",
      "Replaced AASBCR010000046 with GCA_012253045.1_PDT000468022.1_genomic.fna\n",
      "Replaced DABMTO010000031 with GCA_013972605.1_PDT000799260.1_genomic.fna\n",
      "Replaced DAAZWL010000041 with GCA_012473115.1_PDT000205215.2_genomic.fna\n",
      "Replaced DABMTZ010000016 with GCA_013976925.1_PDT000799223.1_genomic.fna\n",
      "Replaced DADKZX010000047 with GCA_017185495.1_PDT000975426.1_genomic.fna\n",
      "Replaced ABACCS010000062 with GCA_018037965.1_PDT000884035.1_genomic.fna\n",
      "Replaced ABCPWX010000041 with GCA_020095175.1_PDT001135985.1_genomic.fna\n",
      "Replaced DABFFC010000036 with GCA_012876535.1_PDT000713244.1_genomic.fna\n",
      "Replaced AASLBN010000047 with GCA_012412525.1_PDT000577098.1_genomic.fna\n",
      "Replaced DABCIR010000055 with GCA_012682085.1_PDT000131779.2_genomic.fna\n",
      "Replaced AASVUK010000050 with GCA_012597925.1_PDT000645056.1_genomic.fna\n",
      "Replaced DABUMK010000061 with GCA_014681525.1_PDT000841049.1_genomic.fna\n",
      "Replaced UWLF01000068 with GCA_900603565.1_E0386_genomic.fna\n",
      "Replaced AATUFM010000085 with GCA_013156485.1_PDT000744528.1_genomic.fna\n",
      "Replaced AASATE010000038 with GCA_012248185.1_PDT000477093.1_genomic.fna\n",
      "Replaced DADHKV010000015 with GCA_017017775.1_PDT000972868.1_genomic.fna\n",
      "Replaced DACABZ010000048 with GCA_015196845.1_PDT000851850.1_genomic.fna\n",
      "Replaced AASFKH010000012 with GCA_012312705.1_PDT000544284.1_genomic.fna\n",
      "Replaced AAXIVO010000051 with GCA_015563045.1_PDT000892312.1_genomic.fna\n",
      "Replaced DABRPB010000036 with GCA_014477235.1_PDT000829549.1_genomic.fna\n",
      "Replaced DABVMW010000067 with GCA_014735055.1_PDT000844076.1_genomic.fna\n",
      "Replaced AASHPN010000039 with GCA_012346175.1_PDT000259583.2_genomic.fna\n",
      "Replaced DABVSN010000048 with GCA_014749985.1_PDT000844125.1_genomic.fna\n",
      "Replaced AASEEH010000094 with GCA_012360155.1_PDT000496910.1_genomic.fna\n",
      "Replaced DADEMK010000061 with GCA_016695825.1_PDT000937717.1_genomic.fna\n",
      "Replaced AATMZF020000088 with GCA_012917825.2_ASM1291782v2_genomic.fna\n",
      "Replaced AAUEYS010000037 with GCA_013629595.1_PDT000773823.1_genomic.fna\n",
      "Replaced AASAZK010000080 with GCA_012251525.1_PDT000479795.1_genomic.fna\n",
      "Replaced DABPSJ010000043 with GCA_014210175.1_PDT000809805.1_genomic.fna\n",
      "Replaced AAWIDC010000048 with GCA_014747845.1_PDT000844503.1_genomic.fna\n",
      "Replaced AASXFA010000068 with GCA_012620585.1_PDT000620262.1_genomic.fna\n",
      "Replaced DABUKV010000037 with GCA_014687025.1_PDT000841197.1_genomic.fna\n",
      "Replaced CVOH01000052 with GCA_001245225.1_E._coli_A192PP_genomic.fna\n",
      "Replaced ABCCPK010000075 with GCA_019752935.1_PDT001116355.1_genomic.fna\n",
      "Replaced DAEPFP010000043 with GCA_018924365.1_PDT001071298.1_genomic.fna\n",
      "Replaced AASXJW010000045 with GCA_012623145.1_PDT000611905.1_genomic.fna\n",
      "Replaced ABAFKJ010000040 with GCA_018122785.1_PDT001010716.1_genomic.fna\n",
      "Replaced AAWOEP010000045 with GCA_014942805.1_PDT000863768.1_genomic.fna\n",
      "Replaced AARZXK010000039 with GCA_012218165.1_PDT000299794.2_genomic.fna\n",
      "Replaced JAHWFJ010000048 with GCA_019311765.1_ASM1931176v1_genomic.fna\n",
      "Replaced DADJCD010000030 with GCA_017075295.1_PDT000951058.1_genomic.fna\n",
      "Replaced AATHSQ010000044 with GCA_012810415.1_PDT000069145.2_genomic.fna\n",
      "Replaced DADUID010000049 with GCA_017807315.1_PDT000998097.1_genomic.fna\n",
      "Replaced DADUIW010000062 with GCA_017815355.1_PDT001000680.1_genomic.fna\n",
      "Replaced AATOND010000020 with GCA_012994205.1_PDT000565931.1_genomic.fna\n",
      "Replaced AAZZYO010000048 with GCA_017953895.1_PDT001005456.1_genomic.fna\n",
      "Replaced AAUNWF010000061 with GCA_013776205.1_PDT000753807.1_genomic.fna\n",
      "Replaced RDDC01000054 with GCA_003686805.1_ASM368680v1_genomic.fna\n",
      "Replaced AAWIEA010000053 with GCA_014747345.1_PDT000844473.1_genomic.fna\n",
      "Replaced CP025403 with GCA_002846135.1_ASM284613v1_genomic.fna\n",
      "Replaced DABZJD010000040 with GCA_015094695.1_PDT000855966.1_genomic.fna\n",
      "Replaced DADFRF010000109 with GCA_016946115.1_PDT000971935.1_genomic.fna\n",
      "['GCA_013039965.1', 'GCA_003794675.1', 'GCA_017753105.1', 'GCA_012609475.1', 'GCA_012375295.1', 'GCA_012602615.1', 'GCA_012567285.1', 'GCA_012898485.1', 'GCA_016108845.1', 'GCA_012359775.1', 'GCA_012375035.1', 'GCA_017016775.1', 'GCA_018207555.1', 'GCA_017253655.1', 'GCA_016387065.1', 'GCA_016108305.1', 'GCA_017924855.1', 'GCA_019215225.1', 'GCA_019883145.1', 'GCA_014485755.1', 'GCA_012045625.1', 'GCA_017947765.1', 'GCA_018180115.1', 'GCA_019002085.1', 'GCA_012790235.1', 'GCA_900448695.1', 'GCA_012298145.1', 'GCA_014822515.1', 'GCA_003784065.1', 'GCA_017675485.1', 'GCA_018837575.1', 'GCA_012355435.1', 'GCA_002513145.1', 'GCA_019385775.1', 'GCA_003069365.1', 'GCA_012865015.1', 'GCA_015621365.1', 'GCA_018162935.1', 'GCA_012973205.1', 'GCA_016903915.1', 'GCA_018258815.1', 'GCA_012989585.1', 'GCA_018560825.1', 'GCA_016387985.1', 'GCA_018162755.1', 'GCA_012416045.1', 'GCA_012534495.1', 'GCA_012279835.1', 'GCA_019297555.1', 'GCA_012469615.1', 'GCA_014736815.1', 'GCA_012625005.1', 'GCA_019736215.1', 'GCA_020091045.1', 'GCA_012383135.1', 'GCA_012224975.1', 'GCA_014751945.1', 'GCA_017786325.1', 'GCA_012358635.1', 'GCA_014461235.1', 'GCA_019366045.1', 'GCA_014990325.1', 'GCA_012990985.1', 'GCA_016698065.1', 'GCA_014809755.1', 'GCA_012470955.1', 'GCA_013016135.1', 'GCA_015484305.1', 'GCA_012621345.1', 'GCA_019885965.1', 'GCA_012343035.1', 'GCA_012314505.1', 'GCA_002930345.1', 'GCA_017047755.1', 'GCA_019577135.1', 'GCA_012576525.1', 'GCA_008033495.1', 'GCA_000397665.1', 'GCA_017031395.1', 'GCA_012253045.1', 'GCA_013972605.1', 'GCA_012473115.1', 'GCA_014745815.1', 'GCA_012251025.2', 'GCA_013038365.1', 'GCA_002087575.1', 'GCA_013976925.1', 'GCA_017185495.1', 'GCA_018037965.1', 'GCA_020095175.1', 'GCA_012876535.1', 'GCA_012412525.1', 'GCA_012682085.1', 'GCA_014478695.1', 'GCA_012597925.1', 'GCA_014681525.1', 'GCA_900603565.1', 'GCA_017045815.1', 'GCA_013156485.1', 'GCA_019483605.1', 'GCA_012248185.1', 'GCA_017017775.1', 'GCA_015196845.1', 'GCA_019075705.1', 'GCA_014343535.1', 'GCA_012312705.1', 'GCA_017947245.1', 'GCA_013050325.1', 'GCA_019850905.1', 'GCA_015563045.1', 'GCA_014477235.1', 'GCA_014735055.1', 'GCA_013038525.1', 'GCA_012346175.1', 'GCA_014749985.1', 'GCA_012360155.1', 'GCA_012608875.1', 'GCA_016695825.1', 'GCA_012917825.2', 'GCA_013629595.1', 'GCA_017753115.1', 'GCA_012993245.1', 'GCA_017758785.1', 'GCA_012251525.1', 'GCA_012622765.1', 'GCA_014210175.1', 'GCA_015163315.1', 'GCA_020095655.1', 'GCA_014747845.1', 'GCA_012620585.1', 'GCA_014687025.1', 'GCA_001245225.1', 'GCA_012833155.1', 'GCA_017785155.1', 'GCA_019752935.1', 'GCA_014076685.2', 'GCA_012071145.1', 'GCA_018924365.1', 'GCA_012623145.1', 'GCA_019885825.1', 'GCA_018122785.1', 'GCA_018290435.1', 'GCA_014942805.1', 'GCA_012218165.1', 'GCA_019311765.1', 'GCA_003772545.1', 'GCA_017075295.1', 'GCA_014081525.1', 'GCA_012810415.1', 'GCA_016107805.1', 'GCA_017807315.1', 'GCA_017815355.1', 'GCA_012994205.1', 'GCA_004960895.1', 'GCA_017953895.1', 'GCA_013776205.1', 'GCA_003686805.1', 'GCA_014747345.1', 'GCA_002846135.1', 'GCA_015094695.1', 'GCA_016946115.1']\n"
     ]
    }
   ],
   "source": [
    "# Change the name of the E. coli genomes of the promoters df to the assembly codes\n",
    "\n",
    "# The input and output paths\n",
    "pathPromoters = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/promoters_df.csv'\n",
    "outputPath = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Results/promoters_w_accession_codes_df.csv'\n",
    "\n",
    "# Read the promoters df\n",
    "promoters = pd.read_csv(pathPromoters, sep='\\t')\n",
    "\n",
    "# Get the original names\n",
    "originalNames = promoters['Unnamed: 0'].tolist()\n",
    "\n",
    "# Replace the names with contig names\n",
    "updatedNames = replace_with_fasta_name(originalNames, pathDirectory)\n",
    "\n",
    "# Replace the names with incomplete assembly codes\n",
    "for i, code in enumerate(updatedNames):\n",
    "    if '.' not in code:\n",
    "        updatedNames[i] = next((item for item in updatedCodes if item.startswith(code)), code)\n",
    "\n",
    "print(updatedNames)\n",
    "\n",
    "# Update the promoters df\n",
    "promoters['Unnamed: 0'] = updatedNames\n",
    "\n",
    "# Save the updated promoters df\n",
    "promoters.to_csv(outputPath, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now let's report in our E. coli df the genomes that contain the iroN gene and which promoters do they have\n",
    "\n",
    "# The input and output paths\n",
    "pathPromoters = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Results/promoters_w_accession_codes_df.csv'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
