{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Eugenio Perez Molphe Montoya ####\n",
    "#### 17.06.2024 ####\n",
    "#### Create graphs to explore what happened ####\n",
    "'''\n",
    "1. An histogram with the percentage identity of the alignments\n",
    "2. The number of genomes that were fed into BProm\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Libraries ###\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import numpy as np\n",
    "from Bio import Entrez\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histogram of percentage identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the percentage identity of the alignments\n",
    "pathDir = '/mnt/mnemo5/eugenio/IroN_project/Files/03_IroN_BLAST_results'\n",
    "blastFiles = os.listdir(pathDir)\n",
    "\n",
    "# Create a list with the percentage identity of the alignments\n",
    "percentageIdentitySalmonella = []\n",
    "percentageIdentityEnterobacteriaceae = []\n",
    "for blastFile in blastFiles:\n",
    "    if blastFile.startswith('GCA_'):\n",
    "        evaluePI = {}\n",
    "        with open(pathDir + '/' + blastFile, 'r') as file:\n",
    "            for line in file:\n",
    "                if line.startswith('#'):\n",
    "                    continue\n",
    "                else:\n",
    "                    line = line.split('\\t')\n",
    "                    evaluePI[float(line[10])] = float(line[2])\n",
    "        if evaluePI:\n",
    "            percentageIdentitySalmonella.append(evaluePI[max(evaluePI.keys())])\n",
    "    else:\n",
    "        evaluePI = {}\n",
    "        with open(pathDir + '/' + blastFile, 'r') as file:\n",
    "            for line in file:\n",
    "                if line.startswith('#'):\n",
    "                    continue\n",
    "                else:\n",
    "                    line = line.split('\\t')\n",
    "                    evaluePI[float(line[10])] = float(line[2])\n",
    "        if evaluePI:\n",
    "            percentageIdentityEnterobacteriaceae.append(evaluePI[max(evaluePI.keys())])\n",
    "\n",
    "print(len(percentageIdentitySalmonella))\n",
    "print(len(percentageIdentityEnterobacteriaceae))\n",
    "print(percentageIdentitySalmonella)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the histogram\n",
    "plt.hist(percentageIdentityEnterobacteriaceae, bins=30, color='#8E394E', alpha=0.7)\n",
    "plt.xlabel('Percentage identity')\n",
    "plt.ylabel('Number of alignments')\n",
    "plt.title('Percentage identity of the alignments (max evalue)')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "mean_value = statistics.mean(percentageIdentitySalmonella)\n",
    "median_value = statistics.median(percentageIdentitySalmonella)\n",
    "mode_value = statistics.mean(percentageIdentitySalmonella)\n",
    "\n",
    "print(mean_value)\n",
    "print(median_value)\n",
    "print(mode_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See if there's relation between sequence length and number of promoters found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's create a numpy array to save our results\n",
    "lengthPromoters = np.zeros((0,2))\n",
    "\n",
    "# Column 0 is length, column 1 is the number of promoters found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathDir = '/mnt/mnemo5/eugenio/IroN_project/Files/06_Promoters/BProm_results_salmonella'\n",
    "Files = os.listdir(pathDir)\n",
    "\n",
    "# Create a list with the percentage identity of the alignments\n",
    "\n",
    "for BPromfile in Files:\n",
    "    with open(pathDir + '/' + BPromfile, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    # Let' obtain the length of the sequence\n",
    "    seqLength = int(lines[1].split(\" \")[-1].strip())\n",
    "\n",
    "    # Let's obtain the number of predicted promoters\n",
    "    startCounting = False\n",
    "    numberPromoters = 0\n",
    "    for i in lines:\n",
    "        if i.startswith(\"For\"):\n",
    "            startCounting = True\n",
    "            continue\n",
    "        if startCounting:\n",
    "            numberPromoters += 1\n",
    "    lengthPromoters = np.vstack([lengthPromoters, np.array([seqLength, numberPromoters])])\n",
    "\n",
    "print(lengthPromoters.shape)\n",
    "print(lengthPromoters[0:10, :])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(lengthPromoters[: , 0], bins=30, color='#8E394E', alpha=0.7)\n",
    "plt.xlabel('Sequence length')\n",
    "plt.ylabel('Number of samples')\n",
    "plt.title('Sequence length of the upstream regions')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "column_to_check = 0\n",
    "unique_numbers = np.unique(lengthPromoters[:, column_to_check])\n",
    "\n",
    "print(\"Unique numbers in column {}: {}\".format(column_to_check, unique_numbers))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### This is weird, all the sequences have the same length, let's check the original fasta file\n",
    "pathEntero = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/enterobacteriaceae_upstream_seqs_300_bp_trimmed.fa'\n",
    "pathSalmo = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/salmonella_upstream_seqs_300_bp_trimmed.fa'\n",
    "\n",
    "seqLengths = []\n",
    "with open(pathEntero, 'r') as file:\n",
    "    lines = file.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            continue\n",
    "        else:\n",
    "            seqLengths.append(len(line.strip()))\n",
    "\n",
    "with open(pathSalmo, 'r') as file:\n",
    "    lines = file.readlines()\n",
    "    for line in lines:\n",
    "        if line.startswith('>'):\n",
    "            continue\n",
    "        else:\n",
    "            seqLengths.append(len(line.strip()))\n",
    "\n",
    "print(np.unique(seqLengths))\n",
    "\n",
    "### So basically BProm is taking the first 80 bp of the sequence and ignoring the rest\n",
    "### BProm: 80\n",
    "### Raw seq length: [126 129 132 152 168 174 186 188 189 197 201 247 253 264 271 279 298 300]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(seqLengths, bins=30, color='#8E394E', alpha=0.7)\n",
    "plt.xlabel('Sequence length')\n",
    "plt.ylabel('Number of samples')\n",
    "plt.title('Raw sequence length of the upstream regions')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Histogram of LDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_LDFs(path, LDFlist):\n",
    "    with open(path, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    ldfValue = lines[4].split(' ')[-1].strip()\n",
    "    LDFlist.append(float(ldfValue))\n",
    "\n",
    "LDFs = []\n",
    "inputDir = '/mnt/mnemo5/eugenio/IroN_project/Files/06_Promoters/BProm_results_enterobacteriaceae'\n",
    "fileList = [inputDir + '/' + file for file in os.listdir(inputDir)]\n",
    "\n",
    "for file in fileList:\n",
    "    read_LDFs(file, LDFs)\n",
    "\n",
    "inputDir = '/mnt/mnemo5/eugenio/IroN_project/Files/06_Promoters/BProm_results_salmonella'\n",
    "fileList = [inputDir + '/' + file for file in os.listdir(inputDir)]\n",
    "\n",
    "for file in fileList:\n",
    "    read_LDFs(file, LDFs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(LDFs, bins=50, color='#8E394E', alpha=0.7)\n",
    "plt.xlabel('LDF score')\n",
    "plt.ylabel('Number of samples')\n",
    "plt.title('LDF scores of the predicted promoters')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Position of predicted promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### For the positions of the predicted promoters\n",
    "def read_positions(path, positionsList):\n",
    "    with open(path, 'r') as file:\n",
    "        fileLines = file.readlines()\n",
    "    # First, I'll filter the lines before the first promoter\n",
    "    startProcessing = False\n",
    "    filteredLines = []\n",
    "    for line in fileLines:\n",
    "        if line.startswith(\"For promoter\"):\n",
    "            startProcessing = True\n",
    "        if startProcessing:\n",
    "            filteredLines.append(line)\n",
    "\n",
    "    # Then, I'll get the promoters\n",
    "    promoters = []\n",
    "    currentPromoter = []\n",
    "    # Iterate over the lines and dump the possible promoters into the currentPromoter list\n",
    "    for line in filteredLines:\n",
    "        if line.startswith(\"For promoter\"):\n",
    "            promoters.append(currentPromoter)\n",
    "            currentPromoter = []\n",
    "            continue\n",
    "        # Skip adding the \"For promoter\" line to the current section, we only care about the identified promoters\n",
    "        currentPromoter.append(line)\n",
    "\n",
    "    # Don't forget to add the last section if it exists\n",
    "    if currentPromoter:\n",
    "        promoters.append(currentPromoter)\n",
    "        # Now, let's get the list of the most possible promoters per each identified promoters\n",
    "    # In other words, the promoters influencing iroN\n",
    "    iroNpromoters = []\n",
    "\n",
    "    for promoter in promoters:\n",
    "        # I'll create a dictionary to store the information: keys as scores and values as the name of the promoter\n",
    "        for oligo in promoter:\n",
    "            oligo = oligo.strip().split(\" \")\n",
    "            if oligo[-1] != 'sites':\n",
    "                try:\n",
    "                    distance = int(oligo[10])\n",
    "                except:\n",
    "                    distance = int(oligo[11])\n",
    "            iroNpromoters.append(distance)\n",
    "    \n",
    "    positionsList += iroNpromoters\n",
    "           \n",
    "    return positionsList\n",
    "\n",
    "positionsList = []\n",
    "inputDir = '/mnt/mnemo5/eugenio/IroN_project/Files/06_Promoters/BProm_results_enterobacteriaceae'\n",
    "fileList = [inputDir + '/' + file for file in os.listdir(inputDir)]\n",
    "\n",
    "for file in fileList:\n",
    "    read_positions(file, positionsList)\n",
    "\n",
    "inputDir = '/mnt/mnemo5/eugenio/IroN_project/Files/06_Promoters/BProm_results_salmonella'\n",
    "fileList = [inputDir + '/' + file for file in os.listdir(inputDir)]\n",
    "\n",
    "for file in fileList:\n",
    "    read_positions(file, positionsList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(positionsList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(positionsList, bins=50, color='#8E394E', alpha=0.7)\n",
    "plt.xlabel('# bases upstream from TSS')\n",
    "plt.ylabel('Number of samples')\n",
    "plt.title('Upstream positions of the predicted promoters')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Check the proportion of the promoters within 40 bp upstream\n",
    "promoters40bp = [i for i in positionsList if i <= 40]\n",
    "print(len(promoters40bp))\n",
    "print(len(positionsList))   \n",
    "print(len(promoters40bp) / len(positionsList))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Number of genomes that were fed to BProm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Functions\n",
    "\n",
    "def fetch_genbank_assembly_from_sample(biosample_codes):\n",
    "    \"\"\"\n",
    "    Get the Genbank accession number based on the NCBI code\n",
    "    \"\"\"\n",
    "    Entrez.email = \"eugenio.perez@mls.uzh.ch\"\n",
    "    genbank_assemblies = {}\n",
    "    failedSearches = []\n",
    "    max_retries = 50\n",
    "    retry_delay = 5\n",
    "    for sample_code in biosample_codes:\n",
    "        attempt = 0\n",
    "        while attempt < max_retries:\n",
    "            try:\n",
    "                # Step 1: Search the Assembly database linked to the BioSample\n",
    "                handle = Entrez.esearch(db=\"assembly\", term=f\"{sample_code}[biosample]\")\n",
    "                record = Entrez.read(handle)\n",
    "                handle.close()\n",
    "\n",
    "                # Check if any assembly IDs were found\n",
    "                if record['IdList']:\n",
    "                    assembly_id = record['IdList'][0]\n",
    "\n",
    "                    # Step 2: Fetch the GenBank assembly accession using the assembly ID\n",
    "                    handle = Entrez.esummary(db=\"assembly\", id=assembly_id, report=\"full\")\n",
    "                    summary_record = Entrez.read(handle)\n",
    "                    handle.close()\n",
    "                    print(summary_record['DocumentSummarySet']['DocumentSummary'][0]['Synonym']['Genbank'])\n",
    "                    genbank_assembly = summary_record['DocumentSummarySet']['DocumentSummary'][0]['Synonym']['Genbank']\n",
    "                    genbank_assemblies[sample_code] = genbank_assembly\n",
    "\n",
    "                    attempt = max_retries\n",
    "                else:\n",
    "                    attempt += 1\n",
    "                    if attempt < max_retries:\n",
    "                        print(f\"No assembly found for BioSample {sample_code}. Retrying in {retry_delay} seconds...\")\n",
    "                        time.sleep(retry_delay)\n",
    "                    else:\n",
    "                        genbank_assemblies[sample_code] = sample_code\n",
    "                        failedSearches.append(sample_code)\n",
    "                        print(f\"No assembly found for BioSample {sample_code}\")\n",
    "\n",
    "            except Exception as e:\n",
    "                attempt += 1\n",
    "                if attempt < max_retries:\n",
    "                    print(f\"Error fetching data for BioSample {sample_code}: {e}. Retrying in {retry_delay} seconds...\")\n",
    "                    time.sleep(retry_delay)\n",
    "                else:\n",
    "                    print(f\"Error fetching data for BioSample {sample_code}: {e}\")\n",
    "                    genbank_assemblies[sample_code] = sample_code\n",
    "                    failedSearches.append(sample_code)\n",
    "\n",
    "    print(f\"Failed searches: {failedSearches}\")\n",
    "    return genbank_assemblies\n",
    "\n",
    "\n",
    "def read_Entero_files(path):\n",
    "    # Read the files with Enterobacteriaceae sequences and outputs a set with their genomes\n",
    "\n",
    "    sampleList = []\n",
    "\n",
    "    with open(path, 'r') as f: # Open the file\n",
    "        listHeaders = f.readlines()\n",
    "    \n",
    "    for line in listHeaders: # Iterate over the lines and get the headers of the sequences\n",
    "        if line.startswith('>'):\n",
    "            sampleCode = line.split('.')[1]\n",
    "            sampleList.append(sampleCode)\n",
    "\n",
    "    sample_accessionCodes = fetch_genbank_assembly_from_sample(sampleList) # Get the accession codes\n",
    "    accessionList = list(sample_accessionCodes.values())\n",
    "\n",
    "    return accessionList\n",
    "\n",
    "def read_Salmo_eColi_files(path):\n",
    "    # Read the files with Salmonella and E. coli sequences and outputs a set with their genomes\n",
    "\n",
    "    sampleList = []\n",
    "\n",
    "    with open(path, 'r') as f: # Open the file\n",
    "        listHeaders = f.readlines()\n",
    "    \n",
    "    for line in listHeaders: # Iterate over the lines and get the headers of the sequences\n",
    "        if line.startswith('>'):\n",
    "            sampleCode = line.strip('_upstream_seq\\n')\n",
    "            sampleCode = sampleCode[1:]\n",
    "\n",
    "            if sampleCode.startswith('GCA_'): # E. coli has assembly codes in them, this will help to only obtain the assembly code and not the rest of the header\n",
    "                sampleCode = sampleCode.split('_')[0] + '_' + sampleCode.split('_')[1]\n",
    "            \n",
    "            sampleList.append(sampleCode)\n",
    "\n",
    "    return sampleList\n",
    "\n",
    "def read_dictionary_for_accession_codes(dictionary):\n",
    "    # So some data sets have sample or contig codes, I need the assembly code, so I need a dictionary to translate the former one into the latter\n",
    "    with open(dictionary, 'r') as f:\n",
    "        translator = f.readlines()\n",
    "    codesList = [l.split(',') for l in translator]\n",
    "    codesDict = {}\n",
    "    for genomes in codesList:\n",
    "        codesDict[genomes[0]] = genomes[1]\n",
    "    \n",
    "    return codesDict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Paths\n",
    "\n",
    "# The Enterobacteriaceae fasta files\n",
    "pathEnteroPlus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/enterobacteriaceae_upstream_seqs_300_bp_trimmed.fa'\n",
    "pathEnteroMinus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/upstream_300_bp_minus_enterobacteriaceae.fa'\n",
    "pathEnteroTotalPlus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/enterobacteriaceae_upstream_seqs_300_bp.fa'\n",
    "pathEnteroTrimmedMinus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/upstream_300_bp_minus_enterobacteriaceae.trimmed.fa'\n",
    "\n",
    "# The salmonella fasta files\n",
    "pathSalmoPlus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/salmonella_upstream_seqs_300_bp_trimmed.fa'\n",
    "pathSalmoMinus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/upstream_300_bp_minus_salmonella.fa'\n",
    "pathSalmoTotalPlus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/salmonella_upstream_seqs_300_bp.fa'\n",
    "pathSalmoTrimmedMinus = '/mnt/mnemo5/eugenio/IroN_project/Files/05_Upstream_seqs/upstream_300_bp_minus_salmonella.trimmed.fa'\n",
    "\n",
    "# The E. coli fasta files\n",
    "pathEColi = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/upstream_300_bp_escherichia.fa'\n",
    "pathEColiTrimmed = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/upstream_300_bp_escherichia.trimmed.fa'\n",
    "\n",
    "# The Salmonella dictionary: contigs - accession numbers\n",
    "pathNCBIcodesSalmo = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Salmonella_header_NCBIcode.csv'\n",
    "\n",
    "# The E. coli dictionary: samples - accession numbers\n",
    "pathNCBIcodesEcoli = \"/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Results/Escherichia_contig-NCBI_codes.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCA_000878475.1\n",
      "GCA_002724035.1\n",
      "GCA_902161885.1\n",
      "GCA_900083565.1\n",
      "GCA_905210975.1\n",
      "GCA_019678825.1\n",
      "GCA_020120175.1\n",
      "GCA_017810085.1\n",
      "GCA_000255535.1\n",
      "GCA_000409245.1\n",
      "GCA_002035265.1\n",
      "GCA_006149135.1\n",
      "GCA_000507385.1\n",
      "GCA_014431485.1\n",
      "GCA_002035315.1\n",
      "GCA_000439255.1\n",
      "GCA_000958225.1\n",
      "GCA_018597905.1\n",
      "GCA_000709535.1\n",
      "GCA_000534035.1\n",
      "GCA_019710705.1\n",
      "GCA_017163885.1\n",
      "GCA_001485355.1\n",
      "GCA_001485375.1\n",
      "GCA_900177785.1\n",
      "GCA_001808015.1\n",
      "GCA_000758305.1\n",
      "GCA_018446855.1\n",
      "GCA_902164705.1\n",
      "GCA_900558415.1\n",
      "GCA_900558775.1\n",
      "GCA_900558065.1\n",
      "GCA_900075745.1\n",
      "GCA_900465085.1\n",
      "GCA_000534175.1\n",
      "GCA_000952405.1\n",
      "GCA_000958725.1\n",
      "GCA_003289325.1\n",
      "GCA_003363215.1\n",
      "GCA_015686035.1\n",
      "GCA_004794965.1\n",
      "GCA_006546545.1\n",
      "GCA_018455485.1\n",
      "GCA_015683035.1\n",
      "GCA_015266255.1\n",
      "GCA_015704035.1\n",
      "GCA_016634105.1\n",
      "GCA_016632585.1\n",
      "GCA_016427985.1\n",
      "GCA_018420435.1\n",
      "GCA_019038895.1\n",
      "GCA_019449085.1\n",
      "GCA_001043635.1\n",
      "GCA_900184035.1\n",
      "GCA_002510255.1\n",
      "GCA_015167415.1\n",
      "GCA_001022855.1\n",
      "GCA_902362195.1\n",
      "GCA_902387465.1\n",
      "GCA_905208965.1\n",
      "GCA_018379085.1\n",
      "GCA_002339925.1\n",
      "GCA_002339275.1\n",
      "GCA_002365575.1\n",
      "GCA_002387975.1\n",
      "GCA_002387405.1\n",
      "GCA_002403795.1\n",
      "GCA_002470595.1\n",
      "GCA_002295215.1\n",
      "GCA_002293915.1\n",
      "GCA_010124885.1\n",
      "GCA_002037625.1\n",
      "GCA_011064805.1\n",
      "GCA_902159485.1\n",
      "GCA_902160195.1\n",
      "GCA_902160515.1\n",
      "GCA_020119035.1\n",
      "GCA_020117115.1\n",
      "GCA_003028515.1\n",
      "GCA_000252995.1\n",
      "GCA_009498575.1\n",
      "GCA_900509765.1\n",
      "GCA_008485645.2\n",
      "GCA_013693605.1\n",
      "GCA_016420845.1\n",
      "GCA_009363175.1\n",
      "GCA_009668265.1\n",
      "GCA_009668125.1\n",
      "GCA_009668235.1\n",
      "GCA_902705915.1\n",
      "GCA_016599695.1\n",
      "GCA_011290365.1\n",
      "GCA_014596845.2\n",
      "GCA_001309035.1\n",
      "GCA_001309295.1\n",
      "GCA_016577585.1\n",
      "GCA_016577545.1\n",
      "GCA_016086815.1\n",
      "GCA_016577305.1\n",
      "GCA_016077095.1\n",
      "GCA_016077115.1\n",
      "GCA_016077265.1\n",
      "GCA_008806235.1\n",
      "GCA_002198775.1\n",
      "GCA_016076795.1\n",
      "GCA_002075445.1\n",
      "GCA_002071735.1\n",
      "GCA_002071775.1\n",
      "GCA_002075575.1\n",
      "GCA_002101505.1\n",
      "GCA_002101515.1\n",
      "GCA_002107915.1\n",
      "GCA_002977895.1\n",
      "GCA_002977575.1\n",
      "GCA_003516125.3\n",
      "GCA_002942305.1\n",
      "GCA_002974475.1\n",
      "GCA_003207335.1\n",
      "GCA_003207255.1\n",
      "GCA_016576885.1\n",
      "GCA_016076595.1\n",
      "GCA_009938705.1\n",
      "GCA_016576485.1\n",
      "GCA_016070875.1\n",
      "GCA_013186285.1\n",
      "GCA_013186245.1\n",
      "GCA_013186165.1\n",
      "GCA_008806095.1\n",
      "GCA_008805175.1\n",
      "GCA_008805195.1\n",
      "GCA_015350575.1\n",
      "GCA_017845295.1\n",
      "GCA_902167975.1\n",
      "GCA_902498885.1\n",
      "GCA_900116465.1\n",
      "GCA_900119515.1\n",
      "GCA_900112805.1\n",
      "GCA_900115965.1\n",
      "GCA_001887675.1\n",
      "GCA_019969905.1\n",
      "GCA_006032975.1\n",
      "GCA_010430625.1\n",
      "GCA_011113815.1\n",
      "GCA_006971165.1\n",
      "GCA_008527795.1\n",
      "GCA_002333625.1\n",
      "GCA_001023395.1\n",
      "GCA_001475395.1\n",
      "GCA_001525095.1\n",
      "GCA_016432165.1\n",
      "GCA_016431365.1\n",
      "GCA_003573465.1\n",
      "GCA_003956145.2\n",
      "GCA_003956105.2\n",
      "GCA_003956125.2\n",
      "GCA_004328445.1\n",
      "GCA_014354245.1\n",
      "GCA_016076115.1\n",
      "GCA_016088375.2\n",
      "GCA_014859035.1\n",
      "GCA_016086895.1\n",
      "GCA_000742345.1\n",
      "GCA_016077195.1\n",
      "GCA_002978495.1\n",
      "GCA_002978385.1\n",
      "GCA_016070675.1\n",
      "GCA_000185685.2\n",
      "GCA_013401715.1\n",
      "GCA_014268045.1\n",
      "GCA_014268055.1\n",
      "GCA_900461045.1\n",
      "GCA_018165835.1\n",
      "GCA_002393245.1\n",
      "GCA_016501955.1\n",
      "GCA_017714855.1\n",
      "GCA_018160875.1\n",
      "GCA_001413115.1\n",
      "GCA_016086435.2\n",
      "GCA_016086475.2\n",
      "GCA_001880845.1\n",
      "GCA_015901105.1\n",
      "GCA_016505255.1\n",
      "GCA_004005085.1\n",
      "GCA_016065775.1\n",
      "GCA_016067175.1\n",
      "GCA_016067195.1\n",
      "GCA_019585185.1\n",
      "GCA_013337065.1\n",
      "GCA_013614655.1\n",
      "GCA_013751655.1\n",
      "GCA_015898445.1\n",
      "GCA_015898305.1\n",
      "GCA_016461655.1\n",
      "GCA_018168455.1\n",
      "GCA_902387725.1\n",
      "GCA_002119365.1\n",
      "GCA_006113225.2\n",
      "GCA_006051555.2\n",
      "GCA_003522735.1\n",
      "GCA_003521435.1\n",
      "GCA_010367945.1\n",
      "GCA_007768505.1\n",
      "GCA_007579005.1\n",
      "GCA_019356415.1\n",
      "GCA_007698165.1\n",
      "GCA_010367765.1\n",
      "GCA_010368225.1\n",
      "GCA_010368585.1\n",
      "GCA_001030185.1\n",
      "GCA_000957725.1\n",
      "GCA_001011475.1\n",
      "GCA_001518035.1\n",
      "GCA_001631565.1\n",
      "GCA_003225895.1\n",
      "GCA_008364345.1\n",
      "GCA_018446235.1\n",
      "GCA_017742775.1\n",
      "GCA_900116535.1\n",
      "GCA_005385285.1\n",
      "GCA_900448695.1\n",
      "GCA_000781295.1\n",
      "GCA_002513275.1\n",
      "GCA_001938625.2\n",
      "GCA_002994915.1\n",
      "GCA_003027665.1\n",
      "GCA_002479525.1\n",
      "GCA_015899165.1\n",
      "GCA_010365605.1\n",
      "GCA_004312225.1\n",
      "GCA_900493215.1\n",
      "GCA_900172715.1\n",
      "GCA_001063525.1\n",
      "GCA_001876675.1\n",
      "GCA_002186895.1\n",
      "GCA_002851335.1\n",
      "GCA_002740705.1\n",
      "GCA_003936505.1\n",
      "GCA_016635835.1\n",
      "GCA_018423085.1\n",
      "GCA_019930485.1\n",
      "GCA_008694045.1\n",
      "GCA_002918575.1\n",
      "GCA_002918465.1\n",
      "GCA_004331585.1\n",
      "GCA_013267145.1\n",
      "GCA_009821375.1\n",
      "GCA_013588055.1\n",
      "GCA_006970665.1\n",
      "GCA_013385855.1\n",
      "GCA_015942525.1\n",
      "GCA_002936565.2\n",
      "GCA_000187705.1\n",
      "GCA_016027675.1\n",
      "GCA_001038925.1\n",
      "GCA_002381145.1\n",
      "GCA_002500185.1\n",
      "GCA_002471025.1\n",
      "GCA_002470585.1\n",
      "GCA_002483855.1\n",
      "GCA_002483945.1\n",
      "GCA_002298265.1\n",
      "GCA_011387015.1\n",
      "Failed searches: []\n",
      "GCA_905210975.1\n",
      "GCA_002152315.1\n",
      "GCA_019428125.1\n",
      "GCA_020116315.1\n",
      "GCA_000255535.1\n",
      "GCA_900100995.1\n",
      "GCA_013358215.1\n",
      "GCA_002035265.1\n",
      "GCA_006149135.1\n",
      "GCA_000507385.1\n",
      "GCA_014431485.1\n",
      "GCA_014338425.3\n",
      "GCA_002035315.1\n",
      "GCA_007694255.1\n",
      "GCA_000280495.2\n",
      "GCA_000439255.1\n",
      "GCA_000300455.4\n",
      "GCA_002211925.1\n",
      "GCA_000957965.1\n",
      "GCA_002161915.1\n",
      "GCA_016430465.1\n",
      "GCA_016431185.1\n",
      "GCA_018438325.1\n",
      "GCA_015857505.1\n",
      "GCA_018338735.1\n",
      "GCA_000463095.2\n",
      "GCA_000709535.1\n",
      "GCA_000691205.1\n",
      "GCA_019710705.1\n",
      "GCA_900775535.1\n",
      "GCA_900509305.1\n",
      "GCA_013169755.1\n",
      "GCA_017163885.1\n",
      "GCA_019053235.1\n",
      "GCA_001485355.1\n",
      "GCA_001485375.1\n",
      "GCA_900177785.1\n",
      "GCA_001808015.1\n",
      "GCA_000758305.1\n",
      "GCA_018446855.1\n",
      "GCA_002923255.1\n",
      "GCA_900538115.1\n",
      "GCA_902162585.1\n",
      "GCA_900558675.1\n",
      "GCA_900558245.1\n",
      "GCA_900557865.1\n",
      "GCA_900076915.1\n",
      "GCA_900077265.1\n",
      "GCA_900077255.1\n",
      "GCA_900076185.1\n",
      "GCA_900076535.1\n",
      "GCA_002378045.1\n",
      "GCA_003428425.1\n",
      "GCA_003965425.1\n",
      "GCA_015684855.1\n",
      "GCA_016616025.1\n",
      "GCA_015238615.1\n",
      "GCA_015476685.1\n",
      "GCA_017498345.1\n",
      "GCA_015704165.1\n",
      "GCA_016428435.1\n",
      "GCA_016428135.1\n",
      "GCA_016428065.1\n",
      "GCA_016428665.1\n",
      "GCA_003263525.1\n",
      "GCA_015135075.1\n",
      "GCA_900184035.1\n",
      "GCA_002510255.1\n",
      "GCA_015167415.1\n",
      "GCA_011800315.1\n",
      "GCA_001316815.2\n",
      "GCA_018599105.1\n",
      "GCA_902387765.1\n",
      "GCA_018379085.1\n",
      "GCA_900079995.3\n",
      "GCA_002339485.1\n",
      "GCA_002340015.1\n",
      "GCA_002339275.1\n",
      "GCA_002354195.1\n",
      "GCA_002366925.1\n",
      "GCA_002403795.1\n",
      "GCA_002499965.1\n",
      "GCA_002295215.1\n",
      "GCA_010124885.1\n",
      "GCA_002037625.1\n",
      "GCA_002033765.1\n",
      "GCA_011064805.1\n",
      "GCA_003063705.1\n",
      "GCA_902160345.1\n",
      "GCA_902160515.1\n",
      "GCA_020117115.1\n",
      "GCA_013590775.1\n",
      "GCA_013730235.1\n",
      "GCA_014779575.1\n",
      "GCA_002923025.1\n",
      "GCA_002923085.1\n",
      "GCA_002923065.1\n",
      "GCA_002923125.1\n",
      "GCA_013337655.1\n",
      "GCA_016771845.1\n",
      "GCA_002922995.1\n",
      "GCA_003028515.1\n",
      "GCA_000252995.1\n",
      "GCA_009935945.1\n",
      "GCA_009498575.1\n",
      "GCA_900774485.1\n",
      "GCA_001969305.1\n",
      "GCA_003196435.1\n",
      "GCA_003196305.1\n",
      "GCA_019730595.1\n",
      "GCA_004342285.1\n",
      "GCA_900978845.1\n",
      "GCA_003864115.1\n",
      "GCA_004522185.1\n",
      "GCA_004136725.1\n",
      "GCA_008485645.2\n",
      "GCA_013693605.1\n",
      "GCA_013138145.1\n",
      "GCA_013693595.1\n",
      "GCA_007910085.1\n",
      "GCA_016420845.1\n",
      "GCA_011684115.1\n",
      "GCA_016599695.1\n",
      "GCA_019049595.1\n",
      "GCA_019049545.1\n",
      "GCA_019049515.1\n",
      "GCA_011290365.1\n",
      "GCA_013816795.1\n",
      "GCA_013733095.1\n",
      "GCA_013781985.1\n",
      "GCA_014596845.2\n",
      "GCA_011765685.1\n",
      "GCA_001971205.1\n",
      "GCA_001971245.1\n",
      "GCA_001308965.1\n",
      "GCA_001309095.1\n",
      "GCA_016577745.1\n",
      "GCA_016577665.1\n",
      "GCA_000974965.1\n",
      "GCA_002093925.1\n",
      "GCA_001277275.1\n",
      "GCA_016077445.1\n",
      "GCA_008806275.1\n",
      "GCA_001735605.1\n",
      "GCA_016077175.1\n",
      "GCA_002075735.1\n",
      "GCA_002107895.1\n",
      "GCA_002199045.1\n",
      "GCA_002199005.1\n",
      "GCA_002978085.1\n",
      "GCA_002978075.1\n",
      "GCA_002977985.1\n",
      "GCA_002977055.1\n",
      "GCA_002976915.1\n",
      "GCA_002976735.1\n",
      "GCA_003516105.2\n",
      "GCA_002993165.1\n",
      "GCA_002976495.1\n",
      "GCA_003177155.1\n",
      "GCA_002975055.1\n",
      "GCA_003207715.1\n",
      "GCA_003207135.1\n",
      "GCA_003515895.1\n",
      "GCA_016070895.1\n",
      "GCA_016070495.1\n",
      "GCA_016071005.1\n",
      "GCA_013186105.1\n",
      "GCA_008805655.1\n",
      "GCA_008805805.1\n",
      "GCA_008805595.1\n",
      "GCA_015353915.1\n",
      "GCA_017845055.1\n",
      "GCA_019144375.1\n",
      "GCA_016412015.1\n",
      "GCA_902167975.1\n",
      "GCA_902498885.1\n",
      "GCA_900111655.1\n",
      "GCA_900116465.1\n",
      "GCA_900119515.1\n",
      "GCA_900112805.1\n",
      "GCA_900115965.1\n",
      "GCA_001887675.1\n",
      "GCA_900168185.1\n",
      "GCA_008330085.1\n",
      "GCA_019969905.1\n",
      "GCA_011342745.1\n",
      "GCA_010124045.1\n",
      "GCA_010638865.1\n",
      "GCA_006852755.1\n",
      "GCA_010498935.1\n",
      "GCA_006094515.1\n",
      "GCA_006096155.1\n",
      "GCA_006032975.1\n",
      "GCA_010430625.1\n",
      "GCA_011113815.1\n",
      "GCA_006971165.1\n",
      "GCA_006114365.1\n",
      "GCA_006051035.1\n",
      "GCA_008527795.1\n",
      "GCA_008531195.1\n",
      "GCA_011531405.1\n",
      "GCA_011618785.1\n",
      "GCA_011520085.1\n",
      "GCA_006599785.1\n",
      "GCA_005481875.1\n",
      "GCA_005656495.1\n",
      "GCA_005841185.1\n",
      "GCA_005633735.1\n",
      "GCA_007872645.1\n",
      "GCA_000966085.1\n",
      "GCA_001023345.1\n",
      "GCA_001631455.1\n",
      "GCA_015681735.1\n",
      "GCA_016431695.1\n",
      "GCA_000012005.1\n",
      "GCA_003573465.1\n",
      "GCA_003956145.2\n",
      "GCA_003956105.2\n",
      "GCA_003956125.2\n",
      "GCA_004328445.1\n",
      "GCA_014354245.1\n",
      "GCA_017493765.1\n",
      "GCA_001276125.1\n",
      "GCA_016505125.1\n",
      "GCA_002978805.1\n",
      "GCA_002978775.1\n",
      "GCA_016076555.1\n",
      "GCA_016086875.1\n",
      "GCA_016086315.1\n",
      "GCA_002978545.1\n",
      "GCA_002978405.2\n",
      "GCA_002978325.1\n",
      "GCA_003858305.1\n",
      "GCA_016076405.1\n",
      "GCA_016077015.1\n",
      "GCA_000185685.2\n",
      "GCA_014268045.1\n",
      "GCA_016771795.1\n",
      "GCA_000155975.1\n",
      "GCA_002794395.1\n",
      "GCA_009830315.1\n",
      "GCA_015893685.1\n",
      "GCA_900461045.1\n",
      "GCA_900446945.1\n",
      "GCA_018165835.1\n",
      "GCA_000783445.2\n",
      "GCA_002393245.1\n",
      "GCA_016505565.1\n",
      "GCA_012955585.1\n",
      "GCA_016501955.1\n",
      "GCA_016501445.1\n",
      "GCA_018421745.2\n",
      "GCA_017714855.1\n",
      "GCA_017955015.1\n",
      "GCA_018160855.1\n",
      "GCA_018160875.1\n",
      "GCA_904425445.1\n",
      "GCA_001412725.1\n",
      "GCA_001413115.1\n",
      "GCA_016086435.2\n",
      "GCA_016086475.2\n",
      "GCA_001689745.1\n",
      "GCA_001880845.1\n",
      "GCA_016075055.1\n",
      "GCA_016074955.1\n",
      "GCA_015901105.1\n",
      "GCA_017957935.1\n",
      "GCA_016505255.1\n",
      "GCA_016504925.1\n",
      "GCA_004005085.1\n",
      "GCA_011022025.1\n",
      "GCA_016065775.1\n",
      "GCA_016067175.1\n",
      "GCA_016067195.1\n",
      "GCA_019585185.1\n",
      "GCA_015947645.1\n",
      "GCA_016502115.1\n",
      "GCA_013337065.1\n",
      "GCA_015230695.1\n",
      "GCA_015231155.1\n",
      "GCA_015929465.1\n",
      "GCA_013614655.1\n",
      "GCA_013743055.1\n",
      "GCA_013751895.1\n",
      "GCA_013751655.1\n",
      "GCA_013604545.1\n",
      "GCA_015929735.1\n",
      "GCA_013874155.1\n",
      "GCA_015894005.1\n",
      "GCA_015929505.3\n",
      "GCA_014802845.1\n",
      "GCA_015898445.1\n",
      "GCA_015897865.1\n",
      "GCA_015898305.1\n",
      "GCA_016501915.1\n",
      "GCA_016461655.1\n",
      "GCA_018168455.1\n",
      "GCA_902387725.1\n",
      "GCA_009621955.1\n",
      "GCA_002035475.1\n",
      "GCA_002035285.1\n",
      "GCA_010368345.1\n",
      "GCA_010368965.1\n",
      "GCA_002119365.1\n",
      "GCA_006051015.2\n",
      "GCA_006113225.2\n",
      "GCA_006051555.2\n",
      "GCA_003448395.1\n",
      "GCA_003497195.1\n",
      "GCA_003521525.1\n",
      "GCA_003522735.1\n",
      "GCA_003521435.1\n",
      "GCA_010367945.1\n",
      "GCA_007497745.1\n",
      "GCA_007768505.1\n",
      "GCA_007496825.1\n",
      "GCA_007579005.1\n",
      "GCA_007644995.1\n",
      "GCA_007501425.1\n",
      "GCA_007501705.1\n",
      "GCA_019356415.1\n",
      "GCA_007698165.1\n",
      "GCA_010367725.1\n",
      "GCA_010368685.1\n",
      "GCA_010367765.1\n",
      "GCA_010367885.1\n",
      "GCA_010368025.1\n",
      "GCA_010368085.1\n",
      "GCA_010368465.1\n",
      "GCA_010368385.1\n",
      "GCA_010368225.1\n",
      "GCA_010368585.1\n",
      "GCA_010368425.1\n",
      "GCA_010368665.1\n",
      "GCA_016653555.1\n",
      "GCA_001011455.1\n",
      "GCA_001011175.1\n",
      "GCA_002796425.1\n",
      "GCA_001518435.1\n",
      "GCA_900116535.1\n",
      "GCA_009363135.1\n",
      "GCA_016668015.1\n",
      "GCA_006236115.1\n",
      "GCA_900083945.1\n",
      "GCA_900083975.1\n",
      "GCA_902163545.1\n",
      "GCA_018163515.1\n",
      "GCA_018924245.1\n",
      "GCA_016088085.1\n",
      "GCA_016086375.1\n",
      "GCA_002479525.1\n",
      "GCA_020115535.1\n",
      "GCA_015899165.1\n",
      "GCA_010365605.1\n",
      "GCA_013826985.1\n",
      "GCA_016903715.1\n",
      "GCA_900084245.1\n",
      "GCA_900451405.1\n",
      "GCA_900514705.1\n",
      "GCA_900173595.1\n",
      "GCA_001060275.1\n",
      "GCA_003664915.1\n",
      "GCA_009387915.1\n",
      "GCA_008370605.1\n",
      "GCA_013392195.1\n",
      "GCA_016762535.1\n",
      "GCA_015278055.1\n",
      "GCA_019915345.1\n",
      "GCA_900452035.1\n",
      "GCA_902706515.1\n",
      "GCA_902706475.1\n",
      "GCA_002918575.1\n",
      "GCA_002918465.1\n",
      "GCA_016505395.1\n",
      "GCA_004331585.1\n",
      "GCA_009648935.1\n",
      "GCA_013267145.1\n",
      "GCA_009821375.1\n",
      "GCA_019270455.1\n",
      "GCA_006861365.1\n",
      "GCA_013588055.1\n",
      "GCA_004146135.1\n",
      "GCA_015943365.1\n",
      "GCA_015942525.1\n",
      "GCA_001611055.1\n",
      "GCA_000187705.1\n",
      "GCA_016027675.1\n",
      "GCA_002381145.1\n",
      "GCA_002500185.1\n",
      "GCA_002471025.1\n",
      "GCA_002470585.1\n",
      "GCA_002383125.1\n",
      "GCA_002294065.1\n",
      "GCA_002483855.1\n",
      "GCA_002483945.1\n",
      "GCA_002298265.1\n",
      "GCA_004295325.1\n",
      "GCA_008087065.1\n",
      "GCA_011387015.1\n",
      "GCA_014489435.1\n",
      "GCA_020099255.1\n",
      "Failed searches: []\n",
      "GCA_905210975.1\n",
      "GCA_017810085.1\n",
      "GCA_000409245.1\n",
      "GCA_002035265.1\n",
      "GCA_006149135.1\n",
      "GCA_000507385.1\n",
      "GCA_014431485.1\n",
      "GCA_002035315.1\n",
      "GCA_000439255.1\n",
      "GCA_018597905.1\n",
      "GCA_000709535.1\n",
      "GCA_001485355.1\n",
      "GCA_001485375.1\n",
      "GCA_900177785.1\n",
      "GCA_001808015.1\n",
      "GCA_000758305.1\n",
      "GCA_018446855.1\n",
      "GCA_900558065.1\n",
      "GCA_015266255.1\n",
      "GCA_015704035.1\n",
      "GCA_016634105.1\n",
      "GCA_001043635.1\n",
      "GCA_900184035.1\n",
      "GCA_002510255.1\n",
      "GCA_015167415.1\n",
      "GCA_902387465.1\n",
      "GCA_002339925.1\n",
      "GCA_002365575.1\n",
      "GCA_002387975.1\n",
      "GCA_002387405.1\n",
      "GCA_002470595.1\n",
      "GCA_002293915.1\n",
      "GCA_010124885.1\n",
      "GCA_002037625.1\n",
      "GCA_902160515.1\n",
      "GCA_020117115.1\n",
      "GCA_000252995.1\n",
      "GCA_008485645.2\n",
      "GCA_013693605.1\n",
      "GCA_009363175.1\n",
      "GCA_011290365.1\n",
      "GCA_006032975.1\n",
      "GCA_010430625.1\n",
      "GCA_011113815.1\n",
      "GCA_006971165.1\n",
      "GCA_008527795.1\n",
      "GCA_002333625.1\n",
      "GCA_001023395.1\n",
      "GCA_001525095.1\n",
      "GCA_016432165.1\n",
      "GCA_016431365.1\n",
      "GCA_016076115.1\n",
      "GCA_016088375.2\n",
      "GCA_014859035.1\n",
      "GCA_000185685.2\n",
      "GCA_900461045.1\n",
      "GCA_018165835.1\n",
      "GCA_002393245.1\n",
      "GCA_016501955.1\n",
      "GCA_017714855.1\n",
      "GCA_018160875.1\n",
      "Error fetching data for BioSample SAMN02872772: Search Backend failed: NCBI C++ Exception:\n",
      "    Error: TXCLIENT(CException::eUnknown) \"/pubmed_gen/rbuild/version/20240724/entrez/2.19/src/internal/txclient/TxClient.cpp\", line 1045: ncbi::CTxRawClientImpl::readAll() --- Read failed: Unknown Error, peer: 130.14.18.121:7011\n",
      ". Retrying in 5 seconds...\n",
      "GCA_001413115.1\n",
      "GCA_016086435.2\n",
      "GCA_016086475.2\n",
      "GCA_001880845.1\n",
      "GCA_015901105.1\n",
      "GCA_016505255.1\n",
      "GCA_004005085.1\n",
      "GCA_016065775.1\n",
      "GCA_016067175.1\n",
      "GCA_016067195.1\n",
      "GCA_019585185.1\n",
      "GCA_013337065.1\n",
      "GCA_013751655.1\n",
      "GCA_015898445.1\n",
      "GCA_015898305.1\n",
      "GCA_016461655.1\n",
      "GCA_902387725.1\n",
      "GCA_002119365.1\n",
      "GCA_006113225.2\n",
      "GCA_006051555.2\n",
      "GCA_003522735.1\n",
      "GCA_003521435.1\n",
      "GCA_010367945.1\n",
      "GCA_007768505.1\n",
      "GCA_007579005.1\n",
      "GCA_019356415.1\n",
      "GCA_007698165.1\n",
      "GCA_010367765.1\n",
      "GCA_010368225.1\n",
      "GCA_010368585.1\n",
      "GCA_001030185.1\n",
      "GCA_000957725.1\n",
      "GCA_001011475.1\n",
      "GCA_001518035.1\n",
      "GCA_001631565.1\n",
      "GCA_003225895.1\n",
      "GCA_008364345.1\n",
      "GCA_018446235.1\n",
      "GCA_017742775.1\n",
      "GCA_900116535.1\n",
      "GCA_900448695.1\n",
      "GCA_000781295.1\n",
      "GCA_002513275.1\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: NCBI C++ Exception:\n",
      "    Error: TXCLIENT(CException::eUnknown) \"/pubmed_gen/rbuild/version/20240724/entrez/2.19/src/internal/txclient/TxClient.cpp\", line 1045: ncbi::CTxRawClientImpl::readAll() --- Read failed: Unknown Error, peer: 130.14.18.121:7011\n",
      ". Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "No assembly found for BioSample SAMN06159501. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "No assembly found for BioSample SAMN06159501. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "No assembly found for BioSample SAMN06159501. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN06159501: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "GCA_001938625.2\n",
      "GCA_003027665.1\n",
      "GCA_002479525.1\n",
      "GCA_015899165.1\n",
      "GCA_010365605.1\n",
      "GCA_002918575.1\n",
      "GCA_002918465.1\n",
      "GCA_004331585.1\n",
      "GCA_009821375.1\n",
      "GCA_013588055.1\n",
      "GCA_015942525.1\n",
      "GCA_016027675.1\n",
      "GCA_001038925.1\n",
      "GCA_002381145.1\n",
      "GCA_002500185.1\n",
      "GCA_002471025.1\n",
      "GCA_002470585.1\n",
      "GCA_002483855.1\n",
      "GCA_002483945.1\n",
      "GCA_002298265.1\n",
      "GCA_011387015.1\n",
      "Failed searches: []\n",
      "GCA_019428125.1\n",
      "GCA_900100995.1\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "No assembly found for BioSample SAMN14987543. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN14987543: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "GCA_013358215.1\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "No assembly found for BioSample SAMN03700895. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN03700895: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "GCA_014338425.3\n",
      "GCA_007694255.1\n",
      "GCA_000300455.4\n",
      "GCA_002211925.1\n",
      "GCA_015857505.1\n",
      "GCA_018338735.1\n",
      "GCA_900557865.1\n",
      "GCA_900077255.1\n",
      "GCA_015238615.1\n",
      "GCA_015476685.1\n",
      "GCA_015704165.1\n",
      "GCA_016428135.1\n",
      "GCA_015135075.1\n",
      "GCA_011800315.1\n",
      "GCA_902387765.1\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: NCBI C++ Exception:\n",
      "    Error: TXCLIENT(CException::eUnknown) \"/pubmed_gen/rbuild/version/20240724/entrez/2.19/src/internal/txclient/TxClient.cpp\", line 1045: ncbi::CTxRawClientImpl::readAll() --- Read failed: Unknown Error, peer: 130.14.22.243:7011\n",
      ". Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMEA3935942: Search Backend failed: Database is not supported: assembly. Retrying in 5 seconds...\n",
      "GCA_900079995.3\n",
      "GCA_002339485.1\n",
      "GCA_002340015.1\n",
      "GCA_002354195.1\n",
      "GCA_002366925.1\n",
      "GCA_002499965.1\n",
      "GCA_002033765.1\n",
      "GCA_902160345.1\n",
      "GCA_013730235.1\n",
      "GCA_014779575.1\n",
      "GCA_002923025.1\n",
      "GCA_002923085.1\n",
      "GCA_002923065.1\n",
      "GCA_002923125.1\n",
      "GCA_013337655.1\n",
      "GCA_016771845.1\n",
      "GCA_002922995.1\n",
      "GCA_003864115.1\n",
      "GCA_013138145.1\n",
      "GCA_013693595.1\n",
      "GCA_013816795.1\n",
      "GCA_013733095.1\n",
      "GCA_013781985.1\n",
      "GCA_011342745.1\n",
      "GCA_010124045.1\n",
      "GCA_010638865.1\n",
      "GCA_006852755.1\n",
      "GCA_010498935.1\n",
      "GCA_006094515.1\n",
      "GCA_006096155.1\n",
      "GCA_006114365.1\n",
      "GCA_006051035.1\n",
      "GCA_008531195.1\n",
      "GCA_011531405.1\n",
      "GCA_011618785.1\n",
      "GCA_011520085.1\n",
      "GCA_006599785.1\n",
      "GCA_005481875.1\n",
      "GCA_005656495.1\n",
      "GCA_005841185.1\n",
      "GCA_005633735.1\n",
      "GCA_007872645.1\n",
      "GCA_000966085.1\n",
      "GCA_001631455.1\n",
      "GCA_015681735.1\n",
      "GCA_016431695.1\n",
      "GCA_000012005.1\n",
      "GCA_017493765.1\n",
      "GCA_001276125.1\n",
      "Error fetching data for BioSample SAMN10435582: Search Backend failed: NCBI C++ Exception:\n",
      "    Error: TXCLIENT(CException::eUnknown) \"/pubmed_gen/rbuild/version/20240724/entrez/2.19/src/internal/txclient/TxClient.cpp\", line 1045: ncbi::CTxRawClientImpl::readAll() --- Read failed: Unknown Error, peer: 130.14.18.121:7011\n",
      ". Retrying in 5 seconds...\n",
      "Error fetching data for BioSample SAMN10435582: Search Backend failed: Unable to open connection to #PMQuerySrv_2_0?dbaf=assembly. Retrying in 5 seconds...\n",
      "GCA_016505125.1\n",
      "GCA_014268045.1\n",
      "GCA_000155975.1\n",
      "GCA_009830315.1\n",
      "GCA_900446945.1\n",
      "GCA_000783445.2\n",
      "GCA_016505565.1\n",
      "GCA_012955585.1\n",
      "GCA_016501445.1\n",
      "GCA_018421745.2\n",
      "GCA_017955015.1\n",
      "GCA_018160855.1\n",
      "GCA_904425445.1\n",
      "GCA_001412725.1\n",
      "GCA_001689745.1\n",
      "GCA_016075055.1\n",
      "GCA_016074955.1\n",
      "GCA_017957935.1\n",
      "GCA_016504925.1\n",
      "GCA_011022025.1\n",
      "GCA_015947645.1\n",
      "GCA_016502115.1\n",
      "GCA_015230695.1\n",
      "GCA_015231155.1\n",
      "GCA_015929465.1\n",
      "GCA_013743055.1\n",
      "GCA_013751895.1\n",
      "GCA_013604545.1\n",
      "GCA_015929735.1\n",
      "GCA_013874155.1\n",
      "GCA_015894005.1\n",
      "GCA_015929505.3\n",
      "GCA_014802845.1\n",
      "GCA_015897865.1\n",
      "GCA_016501915.1\n",
      "GCA_009621955.1\n",
      "GCA_002035475.1\n",
      "GCA_002035285.1\n",
      "GCA_010368345.1\n",
      "GCA_010368965.1\n",
      "GCA_006051015.2\n",
      "GCA_003448395.1\n",
      "GCA_003497195.1\n",
      "GCA_003521525.1\n",
      "GCA_007497745.1\n",
      "GCA_007496825.1\n",
      "GCA_007644995.1\n",
      "GCA_007501425.1\n",
      "GCA_007501705.1\n",
      "GCA_010367725.1\n",
      "GCA_010368685.1\n",
      "GCA_010367885.1\n",
      "GCA_010368025.1\n",
      "GCA_010368085.1\n",
      "GCA_010368465.1\n",
      "GCA_010368385.1\n",
      "GCA_010368425.1\n",
      "GCA_010368665.1\n",
      "GCA_016653555.1\n",
      "GCA_001011455.1\n",
      "GCA_001011175.1\n",
      "GCA_002796425.1\n",
      "GCA_009363135.1\n",
      "GCA_006236115.1\n",
      "GCA_900083945.1\n",
      "GCA_900083975.1\n",
      "GCA_902163545.1\n",
      "GCA_018163515.1\n",
      "GCA_018924245.1\n",
      "GCA_016088085.1\n",
      "GCA_016086375.1\n",
      "GCA_020115535.1\n",
      "GCA_013826985.1\n",
      "GCA_016903715.1\n",
      "GCA_900452035.1\n",
      "GCA_016505395.1\n",
      "GCA_009648935.1\n",
      "GCA_019270455.1\n",
      "GCA_006861365.1\n",
      "GCA_015943365.1\n",
      "GCA_002383125.1\n",
      "GCA_002294065.1\n",
      "GCA_004295325.1\n",
      "GCA_008087065.1\n",
      "GCA_014489435.1\n",
      "GCA_020099255.1\n",
      "Failed searches: []\n",
      "Total:  545\n",
      "Plus 401\n",
      "Minus 262\n",
      "With upstream sequences 278\n"
     ]
    }
   ],
   "source": [
    "### Get the number of Enterobacteriaceae genomes\n",
    "\n",
    "# Get the sample codes from the files \n",
    "minus = set(read_Entero_files(pathEnteroMinus))\n",
    "plus = set(read_Entero_files(pathEnteroTotalPlus))\n",
    "trimmedMinus = set(read_Entero_files(pathEnteroTrimmedMinus))\n",
    "trimmedPlus = set(read_Entero_files(pathEnteroPlus))\n",
    "\n",
    "total = plus | minus\n",
    "totalTrimmed = trimmedMinus | trimmedPlus\n",
    "\n",
    "print('Total: ',len(total))\n",
    "print('Plus', len(plus))\n",
    "print('Minus', len(minus))\n",
    "print('With upstream sequences', len(totalTrimmed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total:  1175\n",
      "With upstream sequences 1109\n"
     ]
    }
   ],
   "source": [
    "### Get the number of Escherichia genomes\n",
    "\n",
    "# Read the E. coli contig - accession code dictionary\n",
    "eColiDict = read_dictionary_for_accession_codes(pathNCBIcodesEcoli)\n",
    "\n",
    "# Get the sample codes from the files\n",
    "total = read_Salmo_eColi_files(pathEColi)\n",
    "totalTrimmed = read_Salmo_eColi_files(pathEColiTrimmed)\n",
    "\n",
    "# Now, let's get the accession codes of the ones registered as contigs\n",
    "total = [eColiDict[contig] for contig in total if contig in eColiDict]\n",
    "totalTrimmed = [eColiDict[contig] for contig in totalTrimmed if contig in eColiDict]\n",
    "\n",
    "# Let's transform them into sets\n",
    "total = set(total)\n",
    "totalTrimmed = set(totalTrimmed)\n",
    "\n",
    "print('Total: ',len(total))\n",
    "print('With upstream sequences', len(totalTrimmed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total:  7472\n",
      "Plus 2739\n",
      "Minus 4733\n",
      "With upstream sequences 7471\n"
     ]
    }
   ],
   "source": [
    "### Get the number of Salmonella genomes\n",
    "\n",
    "# Read the salmonella contig - accession code dictionary\n",
    "salmoDict = read_dictionary_for_accession_codes(pathNCBIcodesSalmo)\n",
    "\n",
    "# Get the contig codes from the files\n",
    "minus = read_Salmo_eColi_files(pathSalmoMinus)\n",
    "plus = read_Salmo_eColi_files(pathSalmoTotalPlus)\n",
    "trimmedMinus = read_Salmo_eColi_files(pathSalmoTrimmedMinus)\n",
    "trimmedPlus = read_Salmo_eColi_files(pathSalmoPlus)\n",
    "\n",
    "# Now let's get the accession codes of the ones registered as contigs\n",
    "minus = set([salmoDict[contig] for contig in minus])\n",
    "plus = set([salmoDict[contig] for contig in plus])\n",
    "trimmedMinus = set([salmoDict[contig] for contig in trimmedMinus])\n",
    "trimmedPlus = set([salmoDict[contig] for contig in trimmedPlus])\n",
    "\n",
    "total = plus | minus\n",
    "totalTrimmed = trimmedMinus | trimmedPlus\n",
    "\n",
    "print('Total: ',len(total))\n",
    "print('Plus', len(plus))\n",
    "print('Minus', len(minus))\n",
    "print('With upstream sequences', len(totalTrimmed))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
