{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Eugenio Perez Molphe Montoya ####\n",
    "#### 17.06.2024 ####\n",
    "#### I add the taxonomic information to the genomes that have iroN, then I create figures to visualize the distribution of the gene in Enterobacteriaceae ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In case that I want to reset the kernel\n",
    "\n",
    "%reset -f"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Necessary for everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "import numpy as np\n",
    "import os\n",
    "from matplotlib.colors import ListedColormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read the taxonomic information files ####\n",
    "\n",
    "pathEnterobacteriaceae = '/mnt/mnemo6/nnaepf/for_eugenio/tax_info_complete.txt'\n",
    "\n",
    "# This information will be saved as dictionary with the following structure:\n",
    "# taxonomic identity number : [taxomic identities in the different taxonomic levels]\n",
    "with open(pathEnterobacteriaceae) as f:\n",
    "    enterTaxInfoList = f.readlines()\n",
    "\n",
    "enterTaxInfo = {}\n",
    "for i in enterTaxInfoList:\n",
    "    line = i.split('\\t')\n",
    "    enterTaxInfo[line[0]] = line[1].split(';')\n",
    "    print(enterTaxInfo[line[0]][6])\n",
    "\n",
    "##### To do: try to get the taxonomic information from left to right, instead of right to left ####\n",
    "\n",
    "# enterTaxInfo[x][-2] = species level [7]\n",
    "# enterTaxInfo[x][-3] = genus level [6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Time to read all the files of the assembly directory to get the taxonomic number of each genome\n",
    "directory = '/mnt/mnemo6/nnaepf/Projects/02_enterobac/out/enterobacteriaceae_test/assembly'  # Replace with the actual directory path\n",
    "\n",
    "fastaFilesNames = os.listdir(directory)\n",
    "\n",
    "# Create a dictionary with sample code - taxonomic code\n",
    "genomesDict = {}\n",
    "genomesList = []\n",
    "for i in fastaFilesNames:\n",
    "    if not i.endswith('.fasta'):\n",
    "        continue\n",
    "    else:\n",
    "        sampleCode = i.split('.')[1]\n",
    "        taxonomicCode = i.split('.')[0]\n",
    "        if taxonomicCode in enterTaxInfo:\n",
    "            genomesDict[sampleCode] = taxonomicCode\n",
    "            genomesList.append(taxonomicCode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Read the file with the names of the genomes that were double checked to effectively have iroN, I'll use it to filter out the genomes that we're still doubtful ####\n",
    "\n",
    "pathFilteredPromoters =  '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/filtered_promoters.txt'\n",
    "\n",
    "with open(pathFilteredPromoters, 'r') as f:\n",
    "    promoters = f.readlines()\n",
    "\n",
    "for idx in range(len(promoters)):\n",
    "    promoters[idx] = promoters[idx].strip()\n",
    "\n",
    "# Now, let's get the NCBI code for the Salmonella genomes\n",
    "path = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Salmonella_header_NCBIcode.csv'\n",
    "headerToNCBIcode = {}\n",
    "with open(path, 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    for line in lines:\n",
    "        header, NCBIcode = line.strip().split(',')\n",
    "        header = header.split('.')[0]\n",
    "        headerToNCBIcode[header] = NCBIcode\n",
    "\n",
    "for idx in range(len(promoters)):\n",
    "    promoters[idx] = promoters[idx].split('.')[0]\n",
    "    if promoters[idx] in headerToNCBIcode:\n",
    "        promoters[idx] = headerToNCBIcode[promoters[idx]]\n",
    "\n",
    "# And transform it into a set\n",
    "qcSet = set(promoters)\n",
    "print(promoters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Proportion of IroN in different genera"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Read the iroN taxonomy identity numbers file ####\n",
    "pathEnterobacteriaceae = '/mnt/mnemo5/eugenio/IroN_project/Files/04_Blast_annotations/enterobacteriaceae_all_blast_hists.sample_codes.txt'\n",
    "\n",
    "# Read the file and save it as a list of taxonomic identity numbers\n",
    "with open(pathEnterobacteriaceae) as f:\n",
    "    iroNtaxInfo = f.readlines()\n",
    "\n",
    "for i in range(len(iroNtaxInfo)):\n",
    "    iroNtaxInfo[i] = iroNtaxInfo[i].strip()\n",
    "\n",
    "# Now let's filter out the ones that we're not sure\n",
    "# iroNtaxInfo = [i for i in iroNtaxInfo if i in qcSet]\n",
    "\n",
    "# Now let's replace the sample codes with the taxonomic identity numbers\n",
    "for idx in range(len(iroNtaxInfo)):\n",
    "    iroNtaxInfo[idx] = genomesDict[iroNtaxInfo[idx]]\n",
    "\n",
    "print(len(iroNtaxInfo))\n",
    "\n",
    "# Then I can create a new list with the taxonomic information for each genome that has iroN, I'll count the repeated taxa to get how many genomes with iroN are present in each taxon\n",
    "# And a dictionary: key(taxon) : value(list with two values: genomes iroN with and total genomes)\n",
    "# Then I can create a bar plot with the taxonomic information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Get the proportion of genomes with iroN in general (I'll use only the dataset with Enterobacteriaceae without the extra info)\n",
    "### There were 1158 genomes of iroN, and iroNtaxInfo that contains all the genomes with iroN has 549 elements, meaning that:\n",
    "### 549/1158 = 0.4709, 47.41%\n",
    "print(505/1158 * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create a list of taxonomic information for the genomes that have iroN,\n",
    "# the proportion of genomes with iroN per sp and genus and the number of genomes with iroN ####\n",
    "\n",
    "# The total number of genomes available per sp and genus\n",
    "\n",
    "# Genus\n",
    "genusList = []\n",
    "for i in genomesList: # genomesList has the taxonomic codes of all the genomes (1158)\n",
    "    genusList.append(enterTaxInfo[i][6])\n",
    "\n",
    "countsGenus = {}\n",
    "for string in genusList:\n",
    "    if string in countsGenus:\n",
    "        countsGenus[string] += 1\n",
    "    else:\n",
    "        countsGenus[string] = 1\n",
    "\n",
    "# Species\n",
    "speciesList = []\n",
    "for i in genomesList:\n",
    "    if enterTaxInfo[i][7]:\n",
    "        speciesList.append(enterTaxInfo[i][7].strip())\n",
    "    else:\n",
    "        speciesList.append(enterTaxInfo[i][-1].strip())\n",
    "\n",
    "countsSpecies = {}  \n",
    "for string in speciesList:\n",
    "    if string in countsSpecies:\n",
    "        countsSpecies[string] += 1\n",
    "    else:\n",
    "        countsSpecies[string] = 1\n",
    "\n",
    "# Create the list of taxonomic information for the genomes that have iroN\n",
    "iroNtaxInfoGenus = []\n",
    "iroNtaxInfoSpecies = []\n",
    "for i in iroNtaxInfo:\n",
    "    iroNtaxInfoGenus.append(enterTaxInfo[i][6])\n",
    "    if enterTaxInfo[i][7]:\n",
    "        iroNtaxInfoSpecies.append(enterTaxInfo[i][7].strip())\n",
    "    else:\n",
    "        iroNtaxInfoSpecies.append(enterTaxInfo[i][-1].strip())\n",
    "\n",
    "# Count the number of genomes with iroN per sp and genus\n",
    "countsGenusIroN = {}\n",
    "for string in iroNtaxInfoGenus:\n",
    "    if string in countsGenusIroN:\n",
    "        countsGenusIroN[string] += 1\n",
    "    else:\n",
    "        countsGenusIroN[string] = 1\n",
    "\n",
    "countsSpeciesIroN = {}\n",
    "for string in iroNtaxInfoSpecies:\n",
    "    string = string.strip()\n",
    "    if string in countsSpeciesIroN:\n",
    "        countsSpeciesIroN[string] += 1\n",
    "    else:\n",
    "        countsSpeciesIroN[string] = 1\n",
    "\n",
    "print(countsGenusIroN)\n",
    "print(countsSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### The same but for Salmonella ###\n",
    "\n",
    "# Let's read the list of Salmonella genomes that have iroN\n",
    "pathSalmonella = '/mnt/mnemo5/eugenio/IroN_project/Files/04_Blast_annotations/salmonella_blasted_sequences.ncbi_codes.txt'\n",
    "\n",
    "# Read the file and save it as a list of taxonomic identity numbers\n",
    "with open(pathSalmonella) as f:\n",
    "    salmonellaTaxInfo = f.readlines()\n",
    "salmonellaTaxInfo = [i.strip() for i in salmonellaTaxInfo]\n",
    "print(len(salmonellaTaxInfo))\n",
    "salmonellaTaxInfo = [i for i in salmonellaTaxInfo if i in qcSet]\n",
    "print(len(salmonellaTaxInfo))\n",
    "\n",
    "# Now let's read the metadata file with the taxonomy of the Salmonella species\n",
    "# The path of the metadata file\n",
    "pathMetadata = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/salmonella_metadata.txt'\n",
    "\n",
    "# Read the metadata file\n",
    "with open(pathMetadata, 'r') as f:\n",
    "    metadata = f.readlines()\n",
    "metadata.pop(0) # Remove the header (accession,bioproject_accession,species_pg3,genus_pg3,Completeness,Contamination,number_of_contigs)\n",
    "metadata = [i.strip().split(',') for i in metadata]\n",
    "\n",
    "# Create a dictionary with the taxonomic information of the Salmonella genomes\n",
    "salmonellaTaxInfoDict = {}\n",
    "for i in metadata:\n",
    "    salmonellaTaxInfoDict[i[0]] = [i[2], i[3]]\n",
    "\n",
    "# Create a list of taxonomic information for the total Salmonella genomes with Salmonella\n",
    "speciesListIroN = []\n",
    "for i in salmonellaTaxInfo:\n",
    "    speciesListIroN.append(salmonellaTaxInfoDict[i][0])\n",
    "\n",
    "genusListIroN = []\n",
    "for i in salmonellaTaxInfo:\n",
    "    genusListIroN.append(salmonellaTaxInfoDict[i][1])\n",
    "\n",
    "# And let's count the total number of genomes per sp and genus\n",
    "salmonellaGenusTotal = []\n",
    "for i in metadata:\n",
    "    salmonellaGenusTotal.append(i[3])\n",
    "salmonellaSpTotal = []\n",
    "for i in metadata:\n",
    "    salmonellaSpTotal.append(i[2])\n",
    "\n",
    "# I'll keep using the dictionaries countsGenus and countsSpecies from the last cell\n",
    "for string in salmonellaGenusTotal:\n",
    "    if string in countsGenus:\n",
    "        countsGenus[string] += 1\n",
    "    else:\n",
    "        countsGenus[string] = 1\n",
    "\n",
    "for string in salmonellaSpTotal:\n",
    "    if string in countsSpecies:\n",
    "        countsSpecies[string] += 1\n",
    "    else:\n",
    "        countsSpecies[string] = 1\n",
    "\n",
    "# Count the number of genomes per sp and genus with the same dictionaries countsGenusIroN and countsSpeciesIroN\n",
    "for string in genusListIroN:\n",
    "    if string in countsGenusIroN:\n",
    "        countsGenusIroN[string] += 1\n",
    "    else:\n",
    "        countsGenusIroN[string] = 1\n",
    "\n",
    "for string in speciesListIroN:\n",
    "    if string in countsSpeciesIroN:\n",
    "        countsSpeciesIroN[string] += 1\n",
    "    else:\n",
    "        countsSpeciesIroN[string] = 1\n",
    "\n",
    "print(countsGenusIroN)\n",
    "print(countsSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Final details for the lists ###\n",
    "\n",
    "# So 'unclassified Enterobacteriaceae' and 'Enterobacteriaceae incertae sedis' are the same thing, I'll join them\n",
    "if 'Enterobacteriaceae incertae sedis' in countsGenusIroN:\n",
    "    countsGenusIroN['unclassified Enterobacteriaceae'] += countsGenusIroN['Enterobacteriaceae incertae sedis']\n",
    "    del countsGenusIroN['Enterobacteriaceae incertae sedis']\n",
    "\n",
    "if 'Enterobacteriaceae incertae sedis' in countsGenus:\n",
    "    countsGenus['unclassified Enterobacteriaceae'] += countsGenus['Enterobacteriaceae incertae sedis']\n",
    "    del countsGenus['Enterobacteriaceae incertae sedis']\n",
    "\n",
    "# And let's add the genera that don't have iroN\n",
    "for i in list(countsGenus.keys()):\n",
    "    if i not in list(countsGenusIroN.keys()):\n",
    "        countsGenusIroN[i] = 0\n",
    "\n",
    "# And let's add the species that don't have iroN\n",
    "for i in list(countsSpecies.keys()):\n",
    "    if i not in list(countsSpeciesIroN.keys()):\n",
    "        countsSpeciesIroN[i] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's get the proportion of genus/sp with iroN\n",
    "proportionGenusIroN = {}\n",
    "for i in countsGenusIroN:\n",
    "    proportionGenusIroN[i] = countsGenusIroN[i] / countsGenus[i] * 100\n",
    "\n",
    "proportionSpeciesIroN = {}\n",
    "for i in countsSpeciesIroN:\n",
    "    proportionSpeciesIroN[i] = countsSpeciesIroN[i] / countsSpecies[i] * 100\n",
    "\n",
    "# Let's print the results\n",
    "print(proportionGenusIroN)\n",
    "print(proportionSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Add the update E. coli values\n",
    "'''\n",
    "So I have one file with the list of E. coli genomes with a double checked iroN, and a directory with all the E. coli genomes, I'll calculate the proportion of genomes with iroN in the E. coli genomes directly.\n",
    "The equation used is:\n",
    "number of E. coli genomes with iroN / total number of E. coli genomes * 100\n",
    "The paths are:\n",
    "/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/genomes_w_double_checked_iroN.txt\n",
    "/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/Genomes/\n",
    "Rrespectively\n",
    "'''\n",
    "\n",
    "proportionGenusIroN['Escherichia'] = 1878 / 10643 * 100\n",
    "proportionSpeciesIroN['Escherichia coli'] = 1878 / 10643 * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's graph the results ###\n",
    "\n",
    "# Data for the plot\n",
    "categories = ['unclassified Enterobacteriaceae', 'Rahnella', 'Serratia', 'Yersinia', 'Cedecea', 'Cronobacter', 'Kosakonia', 'Leclercia', 'Enterobacter', \n",
    "              'Lelliottia', 'Pluralibacter', 'Klebsiella/Raoultella group', 'Kluyvera', 'Citrobacter', 'Salmonella', 'Shigella', 'Escherichia']\n",
    "percentages = [proportionGenusIroN[i] for i in categories]\n",
    "\n",
    "# Create the horizontal bar plot\n",
    "plt.figure(figsize=(14, 10))\n",
    "plt.barh(categories, percentages, color='#707070', edgecolor='black')\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Proportion of genomes with iroN gene (%)')\n",
    "plt.ylabel('Genus')\n",
    "plt.title('Proportion of genomes with iroN per genus')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's save the results ###\n",
    "\n",
    "# First let's make sure that the order of the keys of the dictionaries is the same as the one in the categories list\n",
    "countsGenusIroN = {k: countsGenusIroN[k] for k in categories}\n",
    "countsGenus = {k: countsGenus[k] for k in categories}\n",
    "proportionGenusIroN = {k: proportionGenusIroN[k] for k in categories}\n",
    "\n",
    "orderSp = list(countsSpecies.keys())\n",
    "countsSpeciesIroN = {k: countsSpeciesIroN[k] for k in orderSp}\n",
    "countsSpecies = {k: countsSpecies[k] for k in orderSp}\n",
    "proportionSpeciesIroN = {k: proportionSpeciesIroN[k] for k in orderSp}\n",
    "\n",
    "# Data for the pandas dataframe\n",
    "dataGenus = {'Genus': list(countsGenus.keys()), 'Total genomes': list(countsGenus.values()), 'Genomes with iroN': list(countsGenusIroN.values()), 'Proportion (%)': list(proportionGenusIroN.values())}\n",
    "dataSpecies = {'Species': list(countsSpecies.keys()), 'Total genomes': list(countsSpecies.values()), 'Genomes with iroN': list(countsSpeciesIroN.values()), 'Proportion (%)': list(proportionSpeciesIroN.values())}\n",
    "\n",
    "# Create the pandas dataframes\n",
    "dfGenus = pd.DataFrame(dataGenus)\n",
    "dfSpecies = pd.DataFrame(dataSpecies)\n",
    "\n",
    "# Save the dataframes as csv files\n",
    "pathGenus = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Genus_iroN_proportion_w_filtering.csv'\n",
    "pathSp = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Species_iroN_proportion_w_filtering.csv'\n",
    "\n",
    "dfGenus.to_csv(pathGenus, index=False)\n",
    "dfSpecies.to_csv(pathSp, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dfGenus)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I'll do the heatmap of the promoters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now let's read the metadata file with the taxonomy of the Salmonella species\n",
    "# The path of the metadata file\n",
    "pathMetadata = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/salmonella_metadata.txt'\n",
    "\n",
    "# Read the metadata file\n",
    "with open(pathMetadata, 'r') as f:\n",
    "    metadata = f.readlines()\n",
    "metadata.pop(0) # Remove the header (accession,bioproject_accession,species_pg3,genus_pg3,Completeness,Contamination,number_of_contigs)\n",
    "metadata = [i.strip().split(',') for i in metadata]\n",
    "\n",
    "# Create a dictionary with the taxonomic information of the Salmonella genomes\n",
    "salmonellaTaxInfoDict = {}\n",
    "for i in metadata:\n",
    "    salmonellaTaxInfoDict[i[0]] = [i[2], i[3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enteroPath = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/enterobacteriaceae_matrix_all.tsv'\n",
    "salmoPath = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/salmonella_matrix_all.tsv'\n",
    "\n",
    "# Let's read the dataframes\n",
    "dfEntero = pd.read_csv(enteroPath, sep='\\t')\n",
    "dfSalmo = pd.read_csv(salmoPath, sep='\\t')\n",
    "\n",
    "# Let's change the name of the column with the genome names\n",
    "dfEntero = dfEntero.rename(columns={'Unnamed: 0' : 'Genome'})\n",
    "dfSalmo = dfSalmo.rename(columns={'Unnamed: 0' : 'Genome'})\n",
    "\n",
    "dfEntero['Genome'] = dfEntero['Genome'].astype(str)\n",
    "\n",
    "# Let's replace the genome codes with sp names, I need the dictionary enterTaxInfo\n",
    "for i in genomesDict.keys():\n",
    "    dfEntero = dfEntero.replace(i, genomesDict[i])\n",
    "\n",
    "for i in enterTaxInfo:\n",
    "    dfEntero = dfEntero.replace(i, enterTaxInfo[i][7].strip())\n",
    "\n",
    "print(dfEntero)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's do the same for Salmonella\n",
    "pathNCBIcodes = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Salmonella_header_NCBIcode.csv'\n",
    "\n",
    "# Create the dictionary with the NCBI codes\n",
    "with open(pathNCBIcodes, 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    headerToNCBIcodeSalmo = {}\n",
    "    for line in lines:\n",
    "        line = line.strip().split(',')\n",
    "        headerToNCBIcodeSalmo[line[0]] = line[1]\n",
    "\n",
    "# Now let's replace the genome codes with the NCBI taxonomic codes\n",
    "for i in headerToNCBIcodeSalmo:\n",
    "    dfSalmo = dfSalmo.replace(i, headerToNCBIcodeSalmo[i])\n",
    "\n",
    "# And then let's assign the species names to the genomes\n",
    "\n",
    "for i in salmonellaTaxInfoDict:\n",
    "    dfSalmo = dfSalmo.replace(i, salmonellaTaxInfoDict[i][0])\n",
    "\n",
    "print(dfSalmo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's add the Escherichia promoters\n",
    "ecoliPath = '/mnt/mnemo5/eugenio/IroN_project/Files/10_Escherichia/promoters_df_with_iroN.csv'\n",
    "\n",
    "# Read the dataframe\n",
    "dfEscherichia = pd.read_csv(ecoliPath, sep='\\t')\n",
    "dfEscherichia = dfEscherichia.rename(columns={'Unnamed: 0' : 'Genome'})\n",
    "dfEscherichia = dfEscherichia.drop(columns=['microcin', 'mchI', 'mchB'])\n",
    "\n",
    "# Let's add the species name\n",
    "dfEscherichia['Genome'] = 'Escherichia coli'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# So I don't care about promoter's scores and I'll just replace the values with 1\n",
    "dfEntero.replace(2, 1, inplace=True)\n",
    "dfSalmo.replace(2, 1, inplace=True)\n",
    "dfEscherichia.replace(2, 1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### This is to get the percentage of genomes with ArgR/ArgR2 in the Dfs\n",
    "dfSalmo['ArgR-like'] = (dfSalmo['argR'] == 1) | (dfSalmo['argR2'] == 1)\n",
    "dfEntero['ArgR-like'] = (dfEntero['argR'] == 1) | (dfEntero['argR2'] == 1)\n",
    "dfEscherichia['ArgR-like'] = (dfEscherichia['argR'] == 1) | (dfEscherichia['argR2'] == 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Create a dataframe of Salmonella enterica genomes with iroN ####\n",
    "\n",
    "salmonellaInEnteroDF = dfEntero['Genome'] == 'Salmonella enterica'\n",
    "genomesToAdd = dfEntero[salmonellaInEnteroDF]\n",
    "\n",
    "dfSalmo = pd.concat([dfSalmo, genomesToAdd], ignore_index=True)\n",
    "\n",
    "# Then let's remove the genomes that are not Salmonella enterica, i.e. Salmonella bongori\n",
    "dfSalmo.drop(dfSalmo[dfSalmo['Genome'] == 'Salmonella bongori'].index, inplace=True)\n",
    "\n",
    "# Let's fill the NaN values with 0\n",
    "for col in dfSalmo.columns:\n",
    "    if col != 'Genome':\n",
    "        dfSalmo[col] = dfSalmo[col].fillna(0).astype(int)\n",
    "\n",
    "# Remove the columns with only zeroes\n",
    "dfSalmo = dfSalmo.loc[:, (dfSalmo != 0).any(axis=0)]\n",
    "\n",
    "print(dfSalmo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Create a df that contains all Salmonella genomes with iroN ####\n",
    "genusOfInterest = 'Salmonella'\n",
    "dfFiltered = dfEntero[dfEntero['Genome'].str.contains(genusOfInterest)]\n",
    "\n",
    "dfSalmo = pd.concat([dfSalmo, dfFiltered], ignore_index=True)\n",
    "\n",
    "# Let's fill the NaN values with 0\n",
    "for col in dfSalmo.columns:\n",
    "    if col != 'Genome':\n",
    "        dfSalmo[col] = dfSalmo[col].fillna(0).astype(int)\n",
    "\n",
    "dfSalmo = dfSalmo.loc[:, (dfSalmo != 0).any(axis=0)]\n",
    "\n",
    "print(dfSalmo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Create a dataframe with proportions of genomes with iroN per genera ####\n",
    "\n",
    "# Let's correct a Salmonella that escaped the renaming\n",
    "dfSalmo = dfSalmo.replace('CP033348.1', 'Salmonella enterica')\n",
    "dfEntero = dfEntero.replace('Enterobacteriaceae bacterium 9_2_54FAA', 'unclassified')\n",
    "dfEntero = dfEntero.replace('unclassified Citrobacter', 'unclassified')\n",
    "\n",
    "# Remove the genomes of Escherichia coli in dfEntero (I have already those genomes in dfEscherichia)\n",
    "dfEntero = dfEntero[~dfEntero['Genome'].str.contains('Escherichia coli')]\n",
    "\n",
    "# First of all, I'll create a df that contains all the genomes, and then get only the genera as rows\n",
    "dfAll = pd.concat([dfEntero, dfSalmo, dfEscherichia], ignore_index=True)\n",
    "dfAll['Genome'] = dfAll['Genome'].str.split().str[0]\n",
    "\n",
    "# Let's remove the NaNs and force everyone to be an integer\n",
    "for col in dfAll.columns:\n",
    "    if col != 'Genome':\n",
    "        dfAll[col] = dfAll[col].fillna(0).astype(int)\n",
    "\n",
    "dfMergedByGenera = dfAll.groupby('Genome', as_index=False).sum()\n",
    "\n",
    "# Let's create a dictionary that counts the ocurrences of a genus in the df\n",
    "generaCounts = dfAll['Genome'].value_counts()\n",
    "dfMergedByGenera['Total genomes'] = [generaCounts[i] for i in dfMergedByGenera['Genome']]\n",
    "\n",
    "# Time to get the proportion of genomes with certain promoter\n",
    "for col in dfMergedByGenera.columns[1:]:\n",
    "    if col != 'Total genomes':\n",
    "        dfMergedByGenera[col] = dfMergedByGenera[col] / dfMergedByGenera['Total genomes'] * 100\n",
    "\n",
    "# Delete the column total genomes\n",
    "dfMergedByGenera = dfMergedByGenera.drop(columns=['Total genomes', 'argR', 'argR2'])\n",
    "dfMergedByGenera = dfMergedByGenera.rename(columns={'Genome' : 'Genus', 'ArgR-like' : 'argR'})\n",
    "\n",
    "dfMergedByGenera = dfMergedByGenera.round(1)\n",
    "\n",
    "# Remove columns that only have 0\n",
    "dfMergedByGenera = dfMergedByGenera.loc[:, (dfMergedByGenera != 0).any(axis=0)]\n",
    "\n",
    "print(dfMergedByGenera)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Time to create the heatmap ####\n",
    "\n",
    "# Capitalize the first letter of the promoters\n",
    "dfMergedByGenera.columns = [col[0].upper() + col[1:] if col else col for col in dfMergedByGenera.columns]\n",
    "\n",
    "# Let's create a pandas dataframe with the data ordered according to taxonomy, etc\n",
    "df_pivot = dfMergedByGenera.set_index('Genus')\n",
    "\n",
    "genera = ['Escherichia', 'Shigella', 'Salmonella', 'Citrobacter', 'Klebsiella', 'Enterobacter', 'Kosakonia', 'Cedecea', 'Serratia', 'unclassified']\n",
    "df_pivot = df_pivot.reindex(genera)\n",
    "df_pivot.index = df_pivot.index.map(lambda x: 'Klebsiella/Raoultella group' if x == 'Klebsiella' else x)\n",
    "df_pivot = df_pivot.reindex(sorted(df_pivot.columns), axis=1)\n",
    "\n",
    "# Create a mask for zeros\n",
    "mask = df_pivot == 0\n",
    "\n",
    "# Create a color palette based on the results of a HE agar\n",
    "cmap = LinearSegmentedColormap.from_list('mycmap', ['#E7DA92', '#870B17'])\n",
    "cmap = cmap(np.linspace(0, 1, cmap.N))\n",
    "cmap[0] = (1, 1, 1, 0)  # Set the zero to white\n",
    "cmap[-1] = (0, 0, 0, 1)  # Set the last one to black\n",
    "custom_cmap = ListedColormap(cmap)\n",
    "\n",
    "norm = plt.Normalize(vmin=df_pivot.min().min(), vmax=df_pivot.max().max())\n",
    "\n",
    "# Create the figure\n",
    "plt.figure(figsize=(55, 10))\n",
    "sns.heatmap(df_pivot, mask=mask, cmap=custom_cmap, norm=norm, annot=False, fmt='', cbar=True, linewidths=0.1, linecolor='#8B855F')\n",
    "\n",
    "# Rotate the x-axis labels\n",
    "plt.xticks(rotation=0)\n",
    "\n",
    "# Italize the y-axis labels\n",
    "y_labels = df_pivot.index.tolist()\n",
    "for i, label in enumerate(y_labels):\n",
    "    if label == 'unclassified':\n",
    "        plt.gca().get_yticklabels()[i].set_fontstyle('normal')\n",
    "    else:\n",
    "        plt.gca().get_yticklabels()[i].set_fontstyle('italic')\n",
    "\n",
    "# Set the fontsize of the labels\n",
    "fontsize = 24\n",
    "plt.xticks(fontsize=fontsize - 4)\n",
    "plt.yticks(fontsize=fontsize)\n",
    "\n",
    "#### I need to pivot the columns names and find the correct size, order the promoters according to alphabetic order ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's save our promoters dataframe\n",
    "pathPromoters = '/mnt/mnemo5/eugenio/IroN_project/Files/07_Final_Report/Genus_promoters_proportion.csv'\n",
    "df_pivot.to_csv(pathPromoters, index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
