{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Eugenio Perez Molphe Montoya ####\n",
    "#### 17.06.2024 ####\n",
    "#### I add the taxonomic information to the genomes that have iroN, then I create figures to visualize the distribution of the gene in Enterobacteriaceae ####"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing libraries\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read the taxonomic information files ####\n",
    "\n",
    "pathEnterobacteriaceae = '/mnt/mnemo6/nnaepf/for_eugenio/tax_info_complete.txt'\n",
    "\n",
    "# This information will be saved as dictionary with the following structure:\n",
    "# taxonomic identity number : [taxomic identities in the different taxonomic levels]\n",
    "with open(pathEnterobacteriaceae) as f:\n",
    "    enterTaxInfoList = f.readlines()\n",
    "\n",
    "enterTaxInfo = {}\n",
    "for i in enterTaxInfoList:\n",
    "    line = i.split('\\t')\n",
    "    enterTaxInfo[line[0]] = line[1].split(';')\n",
    "    print(enterTaxInfo[line[0]][-3])\n",
    "\n",
    "# enterTaxInfo[x][-2] = species level\n",
    "# enterTaxInfo[x][-3] = genus level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Read the iroN taxonomy identity numbers file ####\n",
    "pathEnterobacteriaceae = '/mnt/mnemo5/eugenio/IroN_project/Files/04_Blast_annotations/enterobacteriaceae_blasted_sequences.taxonomy_identity_numbers.txt'\n",
    "\n",
    "# Read the file and save it as a list of taxonomic identity numbers\n",
    "with open(pathEnterobacteriaceae) as f:\n",
    "    iroNtaxInfo = f.readlines()\n",
    "\n",
    "for i in range(len(iroNtaxInfo)):\n",
    "    iroNtaxInfo[i] = iroNtaxInfo[i].strip()\n",
    "\n",
    "# Let's remove the repeated information, some genomes have multiple hits of iroN,\n",
    "# and that messes up the percentage of genera/sp with iroN\n",
    "iroNtaxInfo = list(set(iroNtaxInfo))\n",
    "print(len(iroNtaxInfo))\n",
    "\n",
    "# Then I can create a new list with the taxonomic information for each genome that has iroN, I'll count the repeated taxa to get how many genomes with iroN are present in each taxon\n",
    "# And a dictionary: key(taxon) : value(list with two values: genomes iroN with and total genomes)\n",
    "# Then I can create a bar plot with the taxonomic information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Create a list of taxonomic information for the genomes that have iroN,\n",
    "# the proportion of genomes with iroN per sp and genus and the number of genomes with iroN ####\n",
    "\n",
    "# The total number of genomes available per sp and genus\n",
    "\n",
    "# Genus\n",
    "genusList = []\n",
    "for i in enterTaxInfo:\n",
    "    genusList.append(enterTaxInfo[i][-3])\n",
    "\n",
    "# I have sp and strange things at this level, I'll standarize everything, so I will only have genera and not sp\n",
    "genusList = [genus.split()[0] for genus in genusList]\n",
    "\n",
    "countsGenus = {}\n",
    "for string in genusList:\n",
    "    if string in countsGenus:\n",
    "        countsGenus[string] += 1\n",
    "    else:\n",
    "        countsGenus[string] = 1\n",
    "\n",
    "# Species\n",
    "speciesList = []\n",
    "for i in enterTaxInfo:\n",
    "    speciesList.append(enterTaxInfo[i][-2])\n",
    "\n",
    "countsSpecies = {}  \n",
    "for string in speciesList:\n",
    "    if string in countsSpecies:\n",
    "        countsSpecies[string] += 1\n",
    "    else:\n",
    "        countsSpecies[string] = 1\n",
    "\n",
    "# Create the list of taxonomic information for the genomes that have iroN\n",
    "iroNtaxInfoGenus = []\n",
    "iroNtaxInfoSpecies = []\n",
    "for i in iroNtaxInfo:\n",
    "    iroNtaxInfoGenus.append(enterTaxInfo[i][-3])\n",
    "    iroNtaxInfoSpecies.append(enterTaxInfo[i][-2])\n",
    "\n",
    "# Now let's make sure that I only have genera in the genus level\n",
    "iroNtaxInfoGenus = [genus.split()[0] for genus in iroNtaxInfoGenus]\n",
    "\n",
    "# Count the number of genomes with iroN per sp and genus\n",
    "countsGenusIroN = {}\n",
    "for string in iroNtaxInfoGenus:\n",
    "    if string in countsGenusIroN:\n",
    "        countsGenusIroN[string] += 1\n",
    "    else:\n",
    "        countsGenusIroN[string] = 1\n",
    "\n",
    "countsSpeciesIroN = {}\n",
    "for string in iroNtaxInfoSpecies:\n",
    "    if string in countsSpeciesIroN:\n",
    "        countsSpeciesIroN[string] += 1\n",
    "    else:\n",
    "        countsSpeciesIroN[string] = 1\n",
    "\n",
    "print(countsGenusIroN)\n",
    "print(countsSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### The same but for Salmonella ###\n",
    "\n",
    "# Let's read the list of Salmonella genomes that have iroN\n",
    "pathSalmonella = ''\n",
    "\n",
    "# Read the file and save it as a list of taxonomic identity numbers\n",
    "with open(pathSalmonella) as f:\n",
    "    salmonellaTaxInfo = f.readlines()\n",
    "salmonellaTaxInfo = [i.strip() for i in salmonellaTaxInfo]\n",
    "print(salmonellaTaxInfo[0:10])\n",
    "\n",
    "# Now let's read the metadata file with the taxonomy of the Salmonella species\n",
    "# The path of the metadata file\n",
    "pathMetadata = ''\n",
    "\n",
    "# Read the metadata file\n",
    "with open(pathMetadata, 'r') as f:\n",
    "    metadata = f.readlines()\n",
    "metadata.pop(0) # Remove the header (accession,bioproject_accession,species_pg3,genus_pg3,Completeness,Contamination,number_of_contigs)\n",
    "metadata = [i.strip().split(',') for i in metadata]\n",
    "\n",
    "# Create a dictionary with the taxonomic information of the Salmonella genomes\n",
    "salmonellaTaxInfoDict = {}\n",
    "for i in metadata:\n",
    "    salmonellaTaxInfoDict[i[0]] = [i[2], i[3]]\n",
    "\n",
    "# Create a list of taxonomic information for the total Salmonella genomes with Salmonella\n",
    "speciesListIroN = []\n",
    "for i in salmonellaTaxInfo:\n",
    "    speciesListIroN.append(salmonellaTaxInfoDict[i][2])\n",
    "\n",
    "genusListIroN = []\n",
    "for i in salmonellaTaxInfo:\n",
    "    genusListIroN.append(salmonellaTaxInfoDict[i][3])\n",
    "\n",
    "# And let's count the total number of genomes per sp and genus\n",
    "salmonellaGenusTotal = []\n",
    "for i in metadata:\n",
    "    salmonellaGenusTotal.append(i[2])\n",
    "salmonellaSpTotal = []\n",
    "for i in metadata:\n",
    "    salmonellaSpTotal.append(i[3])\n",
    "\n",
    "# I'll keep using the dictionaries countsGenus and countsSpecies from the last cell\n",
    "for string in salmonellaGenusTotal:\n",
    "    if string in countsGenus:\n",
    "        countsGenus[string] += 1\n",
    "    else:\n",
    "        countsGenus[string] = 1\n",
    "\n",
    "for string in salmonellaSpTotal:\n",
    "    if string in countsSpecies:\n",
    "        countsSpecies[string] += 1\n",
    "    else:\n",
    "        countsSpecies[string] = 1\n",
    "\n",
    "# Count the number of genomes per sp and genus with the same dictionaries countsGenusIroN and countsSpeciesIroN\n",
    "for string in genusListIroN:\n",
    "    if string in countsGenusIroN:\n",
    "        countsGenusIroN[string] += 1\n",
    "    else:\n",
    "        countsGenusIroN[string] = 1\n",
    "\n",
    "for string in speciesListIroN:\n",
    "    if string in countsSpeciesIroN:\n",
    "        countsSpeciesIroN[string] += 1\n",
    "    else:\n",
    "        countsSpeciesIroN[string] = 1\n",
    "\n",
    "print(countsGenusIroN)\n",
    "print(countsSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Now let's get the proportion of genus/sp with iroN\n",
    "proportionGenusIroN = {}\n",
    "for i in countsGenusIroN:\n",
    "    proportionGenusIroN[i] = countsGenusIroN[i] / countsGenus[i] * 100\n",
    "\n",
    "proportionSpeciesIroN = {}\n",
    "for i in countsSpeciesIroN:\n",
    "    proportionSpeciesIroN[i] = countsSpeciesIroN[i] / countsSpecies[i] * 100\n",
    "\n",
    "# Let's print the results\n",
    "print(proportionGenusIroN)\n",
    "print(proportionSpeciesIroN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's graph the results ###\n",
    "\n",
    "# Data for the plot\n",
    "categories = list(proportionGenusIroN.keys())\n",
    "percentages = list(proportionGenusIroN.values())\n",
    "\n",
    "# Create the horizontal bar plot\n",
    "plt.barh(categories, percentages, color='blue')\n",
    "\n",
    "# Set the labels and title\n",
    "plt.xlabel('Proportion (%)')\n",
    "plt.ylabel('Genus')\n",
    "plt.title('Proportion of Genus with iroN')\n",
    "\n",
    "# Display the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's save the results ###\n",
    "\n",
    "# Data for the pandas dataframe\n",
    "dataGenus = {'Genus': list(countsGenus.keys()), 'Total genomes': list(countsGenus.values()), 'Genomes with iroN': list(countsGenusIroN.values()), 'Proportion (%)': list(proportionGenusIroN.values())}\n",
    "dataSpecies = {'Species': list(countsSpecies.keys()), 'Total genomes': list(countsSpecies.values()), 'Genomes with iroN': list(countsSpeciesIroN.values()), 'Proportion (%)': list(proportionSpeciesIroN.values())}\n",
    "\n",
    "# Create the pandas dataframes\n",
    "dfGenus = pd.DataFrame(dataGenus)\n",
    "dfSpecies = pd.DataFrame(dataSpecies)\n",
    "\n",
    "# Save the dataframes as csv files\n",
    "pathGenus = ''\n",
    "pathSp = ''\n",
    "\n",
    "dfGenus.to_csv(pathGenus, index=False)\n",
    "dfSpecies.to_csv(pathSp, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
